// @midscene/shared/dist/es/mcp/base-tools.mjs
import { parseBase64 } from "@midscene/shared/img";
import { getDebug } from "@midscene/shared/logger";
import { camelToKebab, getKeyAliases } from "../key-alias-utils.mjs";
import { createNamespacedInitArgSchema, extractNamespacedArgs, sanitizeNamespacedArgs } from "./init-arg-utils.mjs";
import { generateCommonTools, generateToolsFromActionSpace } from "./tool-generator.mjs";
function _define_property(obj, key, value) {
    if (key in obj) Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
    });
    else obj[key] = value;
    return obj;
}
const debug = getDebug('mcp:base-tools');
class BaseMidsceneTools {
    getInitArgKeys() {
        return this.initArgSpec ? Object.keys(this.initArgSpec.shape) : [];
    }
    extractAgentInitParam(args) {
        if (!this.initArgSpec) return;
        const extracted = extractNamespacedArgs(args, this.initArgSpec.namespace, this.getInitArgKeys());
        if (this.initArgSpec.adapt) return this.initArgSpec.adapt(extracted);
        return extracted;
    }
    sanitizeToolArgs(args) {
        if (!this.initArgSpec) return args;
        return sanitizeNamespacedArgs(args, this.initArgSpec.namespace, this.getInitArgKeys());
    }
    getAgentInitArgSchema() {
        if (!this.initArgSpec) return {};
        return createNamespacedInitArgSchema(this.initArgSpec.namespace, this.initArgSpec.shape);
    }
    getAgentInitArgCliMetadata() {
        if (!this.initArgSpec?.cli) return;
        const options = Object.fromEntries(this.getInitArgKeys().map((key)=>{
            const canonicalKey = `${this.initArgSpec.namespace}.${key}`;
            const preferredName = this.initArgSpec.cli?.preferredNames?.[key] ?? (this.initArgSpec.cli?.preferBareKeys ? camelToKebab(key) : canonicalKey);
            const acceptedNames = new Set([
                preferredName,
                ...this.initArgSpec.cli?.preferBareKeys ? getKeyAliases(key) : getKeyAliases(canonicalKey)
            ]);
            acceptedNames.delete(preferredName);
            return [
                canonicalKey,
                {
                    preferredName,
                    aliases: [
                        ...acceptedNames
                    ]
                }
            ];
        }));
        return {
            options
        };
    }
    preparePlatformTools() {
        return [];
    }
    async initTools() {
        this.toolDefinitions = [];
        const platformTools = this.preparePlatformTools();
        this.toolDefinitions.push(...platformTools);
        let actionSpace;
        if (this.agent) {
            actionSpace = await this.agent.getActionSpace();
            debug('Action space from agent:', actionSpace.map((a)=>a.name).join(', '));
        } else {
            const tempDevice = this.createTemporaryDevice();
            actionSpace = tempDevice.actionSpace();
            await tempDevice.destroy?.();
            debug('Action space from temporary device:', actionSpace.map((a)=>a.name).join(', '));
        }
        const actionTools = generateToolsFromActionSpace(actionSpace, (args = {})=>this.ensureAgent(this.extractAgentInitParam(args)), (args = {})=>this.sanitizeToolArgs(args), this.getAgentInitArgSchema(), this.getAgentInitArgCliMetadata());
        const commonTools = generateCommonTools((args = {})=>this.ensureAgent(this.extractAgentInitParam(args)), this.getAgentInitArgSchema(), this.getAgentInitArgCliMetadata());
        this.toolDefinitions.push(...actionTools, ...commonTools);
        debug('Total tools prepared:', this.toolDefinitions.length);
    }
    attachToServer(server) {
        this.mcpServer = server;
        if (0 === this.toolDefinitions.length) debug('Warning: No tools to register. Tools may be initialized lazily.');
        for (const toolDef of this.toolDefinitions)this.mcpServer.tool(toolDef.name, toolDef.description, toolDef.schema, toolDef.handler);
        debug('Registered', this.toolDefinitions.length, 'tools');
    }
    async destroy() {
        await this.agent?.destroy?.();
    }
    getToolDefinitions() {
        return this.toolDefinitions;
    }
    setAgent(agent) {
        this.agent = agent;
    }
    buildScreenshotContent(screenshot) {
        const { mimeType, body } = parseBase64(screenshot);
        return [
            {
                type: 'image',
                data: body,
                mimeType
            }
        ];
    }
    buildTextResult(text) {
        return {
            content: [
                {
                    type: 'text',
                    text
                }
            ]
        };
    }
    createDisconnectHandler(platformName) {
        return async ()=>{
            if (!this.agent) return this.buildTextResult('No active connection to disconnect');
            try {
                await this.agent.destroy?.();
            } catch (error) {
                debug('Failed to destroy agent during disconnect:', error);
            }
            this.agent = void 0;
            return this.buildTextResult(`Disconnected from ${platformName}`);
        };
    }
    constructor(){
        _define_property(this, "mcpServer", void 0);
        _define_property(this, "agent", void 0);
        _define_property(this, "toolDefinitions", []);
    }
}
export { BaseMidsceneTools };
