// @midscene/shared/dist/es/mcp/launcher-helper.mjs
function createMCPServerLauncher(config) {
    const { agent, platformName, ToolsManagerClass, MCPServerClass } = config;
    function validateAgent() {
        const device = agent.interface;
        if (!device) throw new Error(`Agent must have an 'interface' property that references the underlying device.
Please ensure your agent instance is properly initialized with a device interface.
Expected: agent.interface to be defined, but got: ${typeof device}
Solution: Check that your agent constructor properly sets the interface property.`);
    }
    function createToolsManager() {
        const toolsManager = new ToolsManagerClass();
        toolsManager.agent = agent;
        return toolsManager;
    }
    function logStartupInfo(mode, additionalInfo) {
        const device = agent.interface;
        console.log(`Starting Midscene ${platformName} MCP Server (${mode})...`);
        console.log(`Agent: ${agent.constructor.name}`);
        console.log(`Device: ${device.constructor.name}`);
        if (additionalInfo?.port !== void 0) console.log(`Port: ${additionalInfo.port}`);
        if (additionalInfo?.host) console.log(`Host: ${additionalInfo.host}`);
    }
    return {
        async launch (options = {}) {
            const { verbose = true } = options;
            validateAgent();
            if (verbose) logStartupInfo('stdio');
            const toolsManager = createToolsManager();
            const server = new MCPServerClass(toolsManager);
            const result = await server.launch();
            if (verbose) console.log(`${platformName} MCP Server started (stdio mode)`);
            return result;
        },
        async launchHttp (options) {
            const { port, host = 'localhost', verbose = true } = options;
            validateAgent();
            if (verbose) logStartupInfo('HTTP', {
                port,
                host
            });
            const toolsManager = createToolsManager();
            const server = new MCPServerClass(toolsManager);
            const result = await server.launchHttp({
                port,
                host
            });
            if (verbose) console.log(`${platformName} MCP Server started on http://${result.host}:${result.port}/mcp`);
            return result;
        }
    };
}
export { createMCPServerLauncher };
