// @midscene/shared/dist/es/env/model-config-manager.mjs
import { decideModelConfigFromIntentConfig } from "./parse-model-config.mjs";
function _define_property(obj, key, value) {
    if (key in obj) Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
    });
    else obj[key] = value;
    return obj;
}
class ModelConfigManager {
    initialize() {
        if (this.isInitialized) return;
        let configMap;
        if (this.modelConfig) {
            this.isolatedMode = true;
            configMap = this.normalizeModelConfig(this.modelConfig);
        } else configMap = this.globalConfigManager?.getAllEnvConfig() || {};
        const defaultConfig = decideModelConfigFromIntentConfig('default', configMap);
        if (!defaultConfig) throw new Error('Model configuration is incomplete: model name (MIDSCENE_MODEL_NAME) is required. See https://midscenejs.com/model-strategy.html');
        const insightConfig = decideModelConfigFromIntentConfig('insight', configMap);
        const planningConfig = decideModelConfigFromIntentConfig('planning', configMap);
        this.modelConfigMap = {
            default: {
                ...defaultConfig,
                createOpenAIClient: this.createOpenAIClientFn
            },
            insight: {
                ...insightConfig || defaultConfig,
                createOpenAIClient: this.createOpenAIClientFn
            },
            planning: {
                ...planningConfig || defaultConfig,
                createOpenAIClient: this.createOpenAIClientFn
            }
        };
        this.isInitialized = true;
    }
    normalizeModelConfig(config) {
        return Object.entries(config).reduce((acc, [key, value])=>{
            if (null == value) return acc;
            acc[key] = String(value);
            return acc;
        }, Object.create(null));
    }
    clearModelConfigMap() {
        if (this.isolatedMode) throw new Error('ModelConfigManager work in isolated mode, so clearModelConfigMap should not be called');
        this.isInitialized = false;
    }
    getModelConfig(intent) {
        if (!this.isInitialized) this.initialize();
        if (!this.modelConfigMap) throw new Error('modelConfigMap is not initialized, which should not happen');
        return this.modelConfigMap[intent];
    }
    getUploadTestServerUrl() {
        const { openaiExtraConfig } = this.getModelConfig('default');
        const serverUrl = openaiExtraConfig?.REPORT_SERVER_URL;
        return serverUrl;
    }
    registerGlobalConfigManager(globalConfigManager) {
        this.globalConfigManager = globalConfigManager;
    }
    throwErrorIfNonVLModel() {
        const modelConfig = this.getModelConfig('default');
        if (!modelConfig.modelFamily) throw new Error('MIDSCENE_MODEL_FAMILY is not set to a visual language model (VL model), the element localization can not be achieved. Check your model configuration. See https://midscenejs.com/model-strategy.html');
    }
    constructor(modelConfig, createOpenAIClientFn){
        _define_property(this, "modelConfigMap", void 0);
        _define_property(this, "isInitialized", false);
        _define_property(this, "isolatedMode", false);
        _define_property(this, "globalConfigManager", void 0);
        _define_property(this, "modelConfig", void 0);
        _define_property(this, "createOpenAIClientFn", void 0);
        this.modelConfig = modelConfig;
        this.createOpenAIClientFn = createOpenAIClientFn;
    }
}
export { ModelConfigManager };
