// @midscene/playground@1.7.5/dist/es/common.mjs
import { findAllMidsceneLocatorField } from "@midscene/core/ai-model";
import { buildDetailedLocateParam } from "@midscene/core/yaml";
const dataExtractionAPIs = [
    'aiQuery',
    'aiBoolean',
    'aiNumber',
    'aiString',
    'aiAsk'
];
const validationAPIs = [
    'aiAssert',
    'aiWaitFor'
];
const noReplayAPIs = [
    ...dataExtractionAPIs,
    ...validationAPIs
];
const formatErrorMessage = (e)=>{
    const errorMessage = e?.message || '';
    if (errorMessage.includes('of different extension')) return 'Conflicting extension detected. Please disable the suspicious plugins and refresh the page. Guide: https://midscenejs.com/quick-experience.html#faq';
    if (errorMessage.includes('NOT_IMPLEMENTED_AS_DESIGNED')) return 'Further actions cannot be performed in the current environment';
    return errorMessage || 'Unknown error';
};
async function parseStructuredParams(action, params, options = {}) {
    if (!action?.paramSchema || !('shape' in action.paramSchema)) return [
        params.prompt || '',
        options
    ];
    const schema = action.paramSchema;
    const keys = schema && 'shape' in schema ? Object.keys(schema.shape) : [];
    const { deviceOptions: _, ...optionsWithoutDeviceOptions } = options;
    const paramObj = {
        ...optionsWithoutDeviceOptions,
        ...options.deviceOptions || {}
    };
    keys.forEach((key)=>{
        if (void 0 !== params[key] && null !== params[key] && '' !== params[key]) paramObj[key] = params[key];
    });
    if (schema) {
        const locatorFieldKeys = findAllMidsceneLocatorField(schema);
        locatorFieldKeys.forEach((locateKey)=>{
            const locatePrompt = params[locateKey];
            if (locatePrompt && 'string' == typeof locatePrompt) {
                const detailedLocateParam = buildDetailedLocateParam(locatePrompt, {
                    deepLocate: options.deepLocate,
                    cacheable: true
                });
                if (detailedLocateParam) paramObj[locateKey] = detailedLocateParam;
            }
        });
    }
    return [
        paramObj
    ];
}
function validateStructuredParams(value, action) {
    if (!value.params) return {
        valid: false,
        errorMessage: 'Parameters are required'
    };
    if (!action?.paramSchema) return {
        valid: true
    };
    try {
        const paramsForValidation = {
            ...value.params
        };
        const schema = action.paramSchema;
        if (schema) {
            const locatorFieldKeys = findAllMidsceneLocatorField(schema);
            locatorFieldKeys.forEach((key)=>{
                if ('string' == typeof paramsForValidation[key]) paramsForValidation[key] = {
                    midscene_location_field_flag: true,
                    prompt: paramsForValidation[key],
                    center: [
                        0,
                        0
                    ],
                    rect: {
                        left: 0,
                        top: 0,
                        width: 0,
                        height: 0
                    }
                };
            });
        }
        action.paramSchema?.parse(paramsForValidation);
    } catch (error) {
        const zodError = error;
        if (zodError.errors && zodError.errors.length > 0) {
            const errorMessages = zodError.errors.filter((err)=>{
                const path = err.path.join('.');
                return !path.includes('center') && !path.includes('rect') && !path.includes('midscene_location_field_flag');
            }).map((err)=>{
                const field = err.path.join('.');
                return `${field}: ${err.message}`;
            });
            if (errorMessages.length > 0) return {
                valid: false,
                errorMessage: `Validation error: ${errorMessages.join(', ')}`
            };
        } else {
            const errorMsg = error instanceof Error ? error.message : 'Unknown validation error';
            return {
                valid: false,
                errorMessage: `Parameter validation failed: ${errorMsg}`
            };
        }
    }
    return {
        valid: true
    };
}
async function executeAction(activeAgent, actionType, actionSpace, value, options) {
    if ('aiAct' !== actionType && void 0 !== options.deepThink) console.warn('[Playground] Received deepThink in non-aiAct action options. deepThink is expected to be used with aiAct during migration.', {
        actionType,
        requestId: options.requestId,
        options
    });
    const action = actionSpace?.find((a)=>a.interfaceAlias === actionType || a.name === actionType);
    if (action && 'function' == typeof activeAgent.callActionInActionSpace) if (value.params) {
        const parsedParams = await parseStructuredParams(action, value.params, options);
        return await activeAgent.callActionInActionSpace(action.name, parsedParams[0]);
    } else {
        const detailedLocateParam = value.prompt ? buildDetailedLocateParam(value.prompt, {
            deepLocate: options.deepLocate,
            cacheable: true
        }) : void 0;
        const { deviceOptions: _, ...optionsWithoutDeviceOptions } = options;
        const actionParams = {
            locate: detailedLocateParam,
            ...optionsWithoutDeviceOptions,
            ...options.deviceOptions || {}
        };
        return await activeAgent.callActionInActionSpace(action.name, actionParams);
    }
    {
        const prompt = value.prompt;
        if ('aiAssert' === actionType) {
            const { pass, thought } = await activeAgent?.aiAssert?.(prompt || '', void 0, {
                keepRawResponse: true,
                ...options
            }) || {};
            return {
                pass: pass || false,
                thought: thought || ''
            };
        }
        if (activeAgent && 'function' == typeof activeAgent[actionType]) return await activeAgent[actionType](prompt, options);
        throw new Error(`Unknown action type: ${actionType}`);
    }
}
export { dataExtractionAPIs, executeAction, formatErrorMessage, noReplayAPIs, parseStructuredParams, validateStructuredParams, validationAPIs };
