// @midscene/core/dist/es/types.mjs
import { existsSync, mkdirSync, readFileSync, rmSync, writeFileSync } from "node:fs";
import { join } from "node:path";
import { restoreImageReferences } from "./dump/screenshot-restoration.mjs";
import { ScreenshotStore } from "./dump/screenshot-store.mjs";
import { ScreenshotItem } from "./screenshot-item.mjs";
export * from "./yaml.mjs";
function _define_property(obj, key, value) {
    if (key in obj) Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
    });
    else obj[key] = value;
    return obj;
}
class UIContext {
}
class ServiceError extends Error {
    constructor(message, dump){
        super(message), _define_property(this, "dump", void 0);
        this.name = 'ServiceError';
        this.dump = dump;
    }
}
function replacerForDumpSerialization(_key, value) {
    if (value && value.constructor?.name === 'Page') return '[Page object]';
    if (value && value.constructor?.name === 'Browser') return '[Browser object]';
    if (value && 'function' == typeof value.toSerializable) return value.toSerializable();
    return value;
}
function reviverForDumpDeserialization(key, value) {
    if ('screenshot' !== key || 'object' != typeof value || null === value) return value;
    ScreenshotItem.isSerialized(value);
    return value;
}
class ExecutionDump {
    serialize(indents) {
        return JSON.stringify(this.toJSON(), replacerForDumpSerialization, indents);
    }
    toJSON() {
        return {
            id: this.id,
            logTime: this.logTime,
            name: this.name,
            description: this.description,
            tasks: this.tasks.map((task)=>({
                    ...task,
                    recorder: task.recorder || []
                })),
            aiActContext: this.aiActContext
        };
    }
    static fromSerializedString(serialized) {
        const parsed = JSON.parse(serialized, reviverForDumpDeserialization);
        return new ExecutionDump(parsed);
    }
    static fromJSON(data) {
        return new ExecutionDump(data);
    }
    collectScreenshots() {
        const screenshots = [];
        for (const task of this.tasks){
            if (task.uiContext?.screenshot instanceof ScreenshotItem) screenshots.push(task.uiContext.screenshot);
            if (task.recorder) {
                for (const record of task.recorder)if (record.screenshot instanceof ScreenshotItem) screenshots.push(record.screenshot);
            }
        }
        return screenshots;
    }
    constructor(data){
        _define_property(this, "id", void 0);
        _define_property(this, "logTime", void 0);
        _define_property(this, "name", void 0);
        _define_property(this, "description", void 0);
        _define_property(this, "tasks", void 0);
        _define_property(this, "aiActContext", void 0);
        this.id = data.id;
        this.logTime = data.logTime;
        this.name = data.name;
        this.description = data.description;
        this.tasks = data.tasks;
        this.aiActContext = data.aiActContext;
    }
}
class ReportActionDump {
    serialize(indents) {
        return JSON.stringify(this.toJSON(), replacerForDumpSerialization, indents);
    }
    serializeWithInlineScreenshots(indents) {
        const processValue = (obj)=>{
            if (obj instanceof ScreenshotItem) return {
                base64: obj.base64,
                capturedAt: obj.capturedAt
            };
            if (Array.isArray(obj)) return obj.map(processValue);
            if (obj && 'object' == typeof obj) {
                const entries = Object.entries(obj).map(([key, value])=>[
                        key,
                        processValue(value)
                    ]);
                return Object.fromEntries(entries);
            }
            return obj;
        };
        const data = processValue(this.toJSON());
        return JSON.stringify(data, null, indents);
    }
    toJSON() {
        return {
            sdkVersion: this.sdkVersion,
            groupName: this.groupName,
            groupDescription: this.groupDescription,
            modelBriefs: this.modelBriefs,
            executions: this.executions.map((exec)=>exec.toJSON()),
            deviceType: this.deviceType
        };
    }
    static fromSerializedString(serialized) {
        const parsed = JSON.parse(serialized, reviverForDumpDeserialization);
        return new ReportActionDump(parsed);
    }
    static fromJSON(data) {
        return new ReportActionDump(data);
    }
    collectAllScreenshots() {
        const screenshots = [];
        for (const execution of this.executions)screenshots.push(...execution.collectScreenshots());
        return screenshots;
    }
    serializeToFiles(basePath) {
        const screenshotsDir = `${basePath}.screenshots`;
        if (!existsSync(screenshotsDir)) mkdirSync(screenshotsDir, {
            recursive: true
        });
        const screenshots = this.collectAllScreenshots();
        for (const screenshot of screenshots){
            const imagePath = join(screenshotsDir, `${screenshot.id}.${screenshot.extension}`);
            if (existsSync(imagePath)) continue;
            const rawBase64 = screenshot.rawBase64;
            writeFileSync(imagePath, Buffer.from(rawBase64, 'base64'));
        }
        writeFileSync(basePath, this.serialize(), 'utf-8');
    }
    static fromFilesAsInlineJson(basePath) {
        const dumpString = readFileSync(basePath, 'utf-8');
        const screenshotsDir = `${basePath}.screenshots`;
        const loadFromExecutionScreenshotDir = (id, mimeType)=>{
            const ext = 'image/jpeg' === mimeType ? 'jpeg' : 'png';
            const filePath = join(screenshotsDir, `${id}.${ext}`);
            if (!existsSync(filePath)) return '';
            const data = readFileSync(filePath);
            return `data:image/${ext};base64,${data.toString('base64')}`;
        };
        const dumpData = JSON.parse(dumpString);
        const store = new ScreenshotStore({
            mode: 'directory',
            reportPath: basePath
        });
        const processedData = restoreImageReferences(dumpData, (ref)=>{
            const executionFileImage = loadFromExecutionScreenshotDir(ref.id, ref.mimeType);
            if (executionFileImage) return executionFileImage;
            if ('inline' === ref.storage) return '';
            return store.loadBase64(ref);
        });
        return JSON.stringify(processedData);
    }
    static cleanupFiles(basePath) {
        const filesToClean = [
            basePath,
            `${basePath}.screenshots`
        ];
        for (const filePath of filesToClean)try {
            rmSync(filePath, {
                force: true,
                recursive: true
            });
        } catch  {}
    }
    static getFilePaths(basePath) {
        return [
            basePath,
            `${basePath}.screenshots`
        ];
    }
    constructor(data){
        _define_property(this, "sdkVersion", void 0);
        _define_property(this, "groupName", void 0);
        _define_property(this, "groupDescription", void 0);
        _define_property(this, "modelBriefs", void 0);
        _define_property(this, "executions", void 0);
        _define_property(this, "deviceType", void 0);
        this.sdkVersion = data.sdkVersion;
        this.groupName = data.groupName;
        this.groupDescription = data.groupDescription;
        this.modelBriefs = data.modelBriefs;
        this.executions = data.executions.map((exec)=>exec instanceof ExecutionDump ? exec : ExecutionDump.fromJSON(exec));
        this.deviceType = data.deviceType;
    }
}
const GroupedActionDump = ReportActionDump;
export { ExecutionDump, GroupedActionDump, ReportActionDump, ServiceError, UIContext };
