// @midscene/web/dist/es/bridge-mode/io-client.mjs
import { assert } from "@midscene/shared/utils";
import { io } from "socket.io-client";
import { BridgeEvent } from "./common.mjs";
function _define_property(obj, key, value) {
    if (key in obj) Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
    });
    else obj[key] = value;
    return obj;
}
class BridgeClient {
    async connect() {
        return new Promise((resolve, reject)=>{
            const forceWebSocket = 'undefined' == typeof XMLHttpRequest;
            this.socket = io(this.endpoint, {
                reconnection: false,
                ...forceWebSocket ? {
                    transports: [
                        'websocket'
                    ]
                } : {},
                query: {
                    version: "1.7.5"
                }
            });
            const timeout = setTimeout(()=>{
                try {
                    this.socket?.offAny();
                    this.socket?.close();
                } catch (e) {
                    console.warn('got error when offing socket', e);
                }
                this.socket = null;
                reject(new Error('failed to connect to bridge server after timeout'));
            }, 5000);
            this.socket.on('disconnect', (reason)=>{
                this.socket = null;
                this.onDisconnect?.();
            });
            this.socket.on('connect_error', (e)=>{
                console.error('bridge-connect-error', e);
                reject(new Error(e || 'bridge connect error'));
            });
            this.socket.on(BridgeEvent.Connected, (payload)=>{
                clearTimeout(timeout);
                this.serverVersion = payload?.version || 'unknown';
                resolve(this.socket);
            });
            this.socket.on(BridgeEvent.Refused, (e)=>{
                console.error('bridge-refused', e);
                try {
                    this.socket?.disconnect();
                } catch (e) {}
                reject(new Error(e || 'bridge refused'));
            });
            this.socket.on(BridgeEvent.Call, (call)=>{
                const id = call.id;
                assert(void 0 !== id, 'call id is required');
                (async ()=>{
                    let response;
                    try {
                        response = await this.onBridgeCall(call.method, call.args);
                    } catch (e) {
                        const errorContent = `Error from bridge client when calling, method: ${call.method}, args: ${call.args}, error: ${e?.message || e}\n${e?.stack || ''}`;
                        console.error(errorContent);
                        return this.socket?.emit(BridgeEvent.CallResponse, {
                            id,
                            error: errorContent
                        });
                    }
                    this.socket?.emit(BridgeEvent.CallResponse, {
                        id,
                        response
                    });
                })();
            });
        });
    }
    disconnect() {
        this.socket?.disconnect();
        this.socket = null;
    }
    constructor(endpoint, onBridgeCall, onDisconnect){
        _define_property(this, "endpoint", void 0);
        _define_property(this, "onBridgeCall", void 0);
        _define_property(this, "onDisconnect", void 0);
        _define_property(this, "socket", void 0);
        _define_property(this, "serverVersion", void 0);
        this.endpoint = endpoint;
        this.onBridgeCall = onBridgeCall;
        this.onDisconnect = onDisconnect;
        this.socket = null;
        this.serverVersion = null;
    }
}
export { BridgeClient };
