// @midscene/core@1.7.5/dist/es/ai-model/service-caller/request-timeout.mjs
const DEFAULT_AI_CALL_TIMEOUT_MS = 180000;
const AI_CALL_HARD_TIMEOUT_CODE = 'AI_CALL_HARD_TIMEOUT';
function resolveEffectiveTimeoutMs(modelConfig) {
    const { timeout } = modelConfig;
    if ('number' != typeof timeout) return DEFAULT_AI_CALL_TIMEOUT_MS;
    if (timeout <= 0) return null;
    return timeout;
}
function isHardTimeoutError(err) {
    if (!err || 'object' != typeof err) return false;
    const code = err.code;
    if (code === AI_CALL_HARD_TIMEOUT_CODE) return true;
    const cause = err.cause;
    if (cause && 'object' == typeof cause && cause.code === AI_CALL_HARD_TIMEOUT_CODE) return true;
    return false;
}
function buildRequestAbortSignal(timeoutMs, userSignal) {
    const controller = new AbortController();
    if (userSignal?.aborted) {
        controller.abort(userSignal.reason);
        return {
            signal: controller.signal,
            cleanup: ()=>{}
        };
    }
    let timer;
    if (null !== timeoutMs) {
        timer = setTimeout(()=>{
            const err = new Error(`AI call hard timeout after ${timeoutMs}ms (full request time exceeded)`);
            err.code = AI_CALL_HARD_TIMEOUT_CODE;
            controller.abort(err);
        }, timeoutMs);
        if ('function' == typeof timer.unref) timer.unref();
    }
    const onUserAbort = userSignal ? ()=>controller.abort(userSignal.reason) : void 0;
    if (userSignal && onUserAbort) userSignal.addEventListener('abort', onUserAbort, {
        once: true
    });
    return {
        signal: controller.signal,
        cleanup: ()=>{
            if (timer) clearTimeout(timer);
            if (userSignal && onUserAbort) userSignal.removeEventListener('abort', onUserAbort);
        }
    };
}
export { AI_CALL_HARD_TIMEOUT_CODE, DEFAULT_AI_CALL_TIMEOUT_MS, buildRequestAbortSignal, isHardTimeoutError, resolveEffectiveTimeoutMs };
