// @midscene/shared@1.7.5/dist/es/env/parse-model-config.mjs
import { DEFAULT_MODEL_CONFIG_KEYS, INSIGHT_MODEL_CONFIG_KEYS, PLANNING_MODEL_CONFIG_KEYS } from "./constants.mjs";
import { MIDSCENE_OPENAI_HTTP_PROXY, MIDSCENE_OPENAI_INIT_CONFIG_JSON, MIDSCENE_OPENAI_SOCKS_PROXY, MIDSCENE_USE_DOUBAO_VISION, MIDSCENE_USE_GEMINI, MIDSCENE_USE_QWEN3_VL, MIDSCENE_USE_QWEN_VL, MIDSCENE_USE_VLM_UI_TARS, MODEL_FAMILY_VALUES, OPENAI_API_KEY, OPENAI_BASE_URL, UITarsModelVersion } from "./types.mjs";
import { getDebug } from "../logger.mjs";
import { assert } from "../utils.mjs";
import { maskConfig, parseJson } from "./helper.mjs";
import { initDebugConfig } from "./init-debug.mjs";
const KEYS_MAP = {
    insight: INSIGHT_MODEL_CONFIG_KEYS,
    planning: PLANNING_MODEL_CONFIG_KEYS,
    default: DEFAULT_MODEL_CONFIG_KEYS
};
const getUITarsModelVersion = (modelFamily)=>{
    if ('vlm-ui-tars' === modelFamily) return UITarsModelVersion.V1_0;
    if ('vlm-ui-tars-doubao' === modelFamily || 'vlm-ui-tars-doubao-1.5' === modelFamily) return UITarsModelVersion.DOUBAO_1_5_20B;
};
const validateModelFamily = (modelFamily)=>{
    if (modelFamily && !MODEL_FAMILY_VALUES.includes(modelFamily)) throw new Error(`Invalid MIDSCENE_MODEL_FAMILY value: ${modelFamily}`);
};
const legacyConfigToModelFamily = (provider)=>{
    const isDoubao = provider[MIDSCENE_USE_DOUBAO_VISION];
    const isQwen = provider[MIDSCENE_USE_QWEN_VL];
    const isQwen3 = provider[MIDSCENE_USE_QWEN3_VL];
    const isUiTars = provider[MIDSCENE_USE_VLM_UI_TARS];
    const isGemini = provider[MIDSCENE_USE_GEMINI];
    const enabledModes = [
        isDoubao && MIDSCENE_USE_DOUBAO_VISION,
        isQwen && MIDSCENE_USE_QWEN_VL,
        isQwen3 && MIDSCENE_USE_QWEN3_VL,
        isUiTars && MIDSCENE_USE_VLM_UI_TARS,
        isGemini && MIDSCENE_USE_GEMINI
    ].filter(Boolean);
    if (enabledModes.length > 1) throw new Error(`Only one vision mode can be enabled at a time. Currently enabled modes: ${enabledModes.join(', ')}. Please disable all but one mode.`);
    if (isQwen3) return 'qwen3-vl';
    if (isQwen) return 'qwen2.5-vl';
    if (isDoubao) return 'doubao-vision';
    if (isGemini) return 'gemini';
    if (isUiTars) if ('1' === isUiTars) return 'vlm-ui-tars';
    else if ('DOUBAO' === isUiTars || 'DOUBAO-1.5' === isUiTars) return 'vlm-ui-tars-doubao-1.5';
    else return 'vlm-ui-tars-doubao';
};
const getModelDescription = (modelFamily, uiTarsModelVersion)=>{
    if (uiTarsModelVersion) return `UI-TARS=${uiTarsModelVersion}`;
    if (modelFamily) return `${modelFamily} mode`;
    return '';
};
const normalizeOpenaiExtraConfig = (config)=>{
    if (!config || 'object' != typeof config || Array.isArray(config)) return;
    const { defaultHeaders, extra_headers, extraHeaders, ...rest } = config;
    const headers = defaultHeaders ?? extra_headers ?? extraHeaders;
    if (void 0 !== headers) return {
        ...rest,
        defaultHeaders: headers
    };
    return rest;
};
const parseOpenaiSdkConfig = ({ keys, provider, useLegacyLogic = false })=>{
    initDebugConfig();
    const debugLog = getDebug('ai:config');
    debugLog('enter parseOpenaiSdkConfig with keys:', keys);
    const legacyAPIKey = useLegacyLogic ? provider[OPENAI_API_KEY] : void 0;
    const legacyBaseURL = useLegacyLogic ? provider[OPENAI_BASE_URL] : void 0;
    const legacySocksProxy = useLegacyLogic ? provider[MIDSCENE_OPENAI_SOCKS_PROXY] : void 0;
    const legacyHttpProxy = useLegacyLogic ? provider[MIDSCENE_OPENAI_HTTP_PROXY] : void 0;
    const legacyOpenaiExtraConfig = useLegacyLogic ? provider[MIDSCENE_OPENAI_INIT_CONFIG_JSON] : void 0;
    const legacyModelFamily = useLegacyLogic ? legacyConfigToModelFamily(provider) : void 0;
    const modelFamilyRaw = provider[keys.modelFamily] || legacyModelFamily;
    const openaiApiKey = provider[keys.openaiApiKey] || legacyAPIKey;
    const openaiBaseURL = provider[keys.openaiBaseURL] || legacyBaseURL;
    const socksProxy = provider[keys.socksProxy] || legacySocksProxy;
    const httpProxy = provider[keys.httpProxy] || legacyHttpProxy;
    const modelName = provider[keys.modelName];
    const openaiExtraConfigStr = provider[keys.openaiExtraConfig];
    const openaiExtraConfig = parseJson(keys.openaiExtraConfig, openaiExtraConfigStr || legacyOpenaiExtraConfig);
    const extraBodyStr = provider[keys.extraBody];
    const extraBody = parseJson(keys.extraBody, extraBodyStr);
    const temperature = provider[keys.temperature] ? Number(provider[keys.temperature]) : 0;
    const modelFamily = modelFamilyRaw;
    validateModelFamily(modelFamily);
    const uiTarsModelVersion = getUITarsModelVersion(modelFamily);
    const modelDescription = getModelDescription(modelFamily, uiTarsModelVersion);
    return {
        socksProxy,
        httpProxy,
        openaiBaseURL,
        openaiApiKey,
        openaiExtraConfig: normalizeOpenaiExtraConfig(openaiExtraConfig),
        extraBody,
        modelFamily,
        uiTarsModelVersion,
        modelName: modelName,
        modelDescription,
        intent: '-',
        timeout: provider[keys.timeout] ? Number(provider[keys.timeout]) : void 0,
        temperature,
        retryCount: (()=>{
            if (!provider[keys.retryCount]) return 1;
            const val = Number(provider[keys.retryCount]);
            if (!Number.isFinite(val)) return 1;
            if (val < 0) throw new Error(`${keys.retryCount} must be non-negative, got ${val}`);
            return val;
        })(),
        retryInterval: (()=>{
            if (!provider[keys.retryInterval]) return 2000;
            const val = Number(provider[keys.retryInterval]);
            if (!Number.isFinite(val)) return 2000;
            if (val < 0) throw new Error(`${keys.retryInterval} must be non-negative, got ${val}`);
            return val;
        })(),
        reasoningEffort: provider[keys.reasoningEffort]?.trim() || void 0,
        reasoningEnabled: (()=>{
            const val = provider[keys.reasoningEnabled]?.trim()?.toLowerCase();
            if ('true' === val || '1' === val) return true;
            if ('false' === val || '0' === val) return false;
        })(),
        reasoningBudget: (()=>{
            const val = provider[keys.reasoningBudget]?.trim();
            if (!val) return;
            const num = Number(val);
            return Number.isFinite(num) ? num : void 0;
        })()
    };
};
const decideModelConfigFromIntentConfig = (intent, configMap)=>{
    const debugLog = getDebug('ai:config');
    debugLog('will decideModelConfig base on agent.modelConfig()', intent, maskConfig(configMap));
    const keysForFn = KEYS_MAP[intent];
    const modelName = configMap[keysForFn.modelName];
    if (!modelName) return void debugLog('no modelName found for intent', intent);
    const finalResult = parseOpenaiSdkConfig({
        keys: keysForFn,
        provider: configMap,
        useLegacyLogic: 'default' === intent
    });
    finalResult.intent = intent;
    debugLog('decideModelConfig result by agent.modelConfig() with intent', intent, maskConfig({
        ...finalResult
    }));
    assert(finalResult.openaiBaseURL, `failed to get base URL of model (intent=${intent}). See https://midscenejs.com/model-strategy`);
    if (!finalResult.modelName) console.warn(`modelName is not set for intent ${intent}, this may cause unexpected behavior. See https://midscenejs.com/model-strategy`);
    return finalResult;
};
export { decideModelConfigFromIntentConfig, getUITarsModelVersion, legacyConfigToModelFamily, parseOpenaiSdkConfig, validateModelFamily };
