// @midscene/shared@1.7.5/dist/es/env/types.mjs
const MIDSCENE_MODEL_INIT_CONFIG_JSON = 'MIDSCENE_MODEL_INIT_CONFIG_JSON';
const MIDSCENE_MODEL_EXTRA_BODY_JSON = 'MIDSCENE_MODEL_EXTRA_BODY_JSON';
const MIDSCENE_MODEL_NAME = 'MIDSCENE_MODEL_NAME';
const MIDSCENE_DEBUG_MODEL_PROFILE = 'MIDSCENE_DEBUG_MODEL_PROFILE';
const MIDSCENE_DEBUG_MODEL_RESPONSE = 'MIDSCENE_DEBUG_MODEL_RESPONSE';
const MIDSCENE_DANGEROUSLY_PRINT_ALL_CONFIG = 'MIDSCENE_DANGEROUSLY_PRINT_ALL_CONFIG';
const MIDSCENE_DEBUG_MODE = 'MIDSCENE_DEBUG_MODE';
const MIDSCENE_MCP_USE_PUPPETEER_MODE = 'MIDSCENE_MCP_USE_PUPPETEER_MODE';
const MIDSCENE_MCP_CHROME_PATH = 'MIDSCENE_MCP_CHROME_PATH';
const MIDSCENE_MCP_ANDROID_MODE = 'MIDSCENE_MCP_ANDROID_MODE';
const DOCKER_CONTAINER = 'DOCKER_CONTAINER';
const MIDSCENE_LANGSMITH_DEBUG = 'MIDSCENE_LANGSMITH_DEBUG';
const MIDSCENE_LANGFUSE_DEBUG = 'MIDSCENE_LANGFUSE_DEBUG';
const MIDSCENE_MODEL_SOCKS_PROXY = 'MIDSCENE_MODEL_SOCKS_PROXY';
const MIDSCENE_MODEL_HTTP_PROXY = 'MIDSCENE_MODEL_HTTP_PROXY';
const MIDSCENE_MODEL_API_KEY = 'MIDSCENE_MODEL_API_KEY';
const MIDSCENE_MODEL_BASE_URL = 'MIDSCENE_MODEL_BASE_URL';
const MIDSCENE_MODEL_MAX_TOKENS = 'MIDSCENE_MODEL_MAX_TOKENS';
const MIDSCENE_MODEL_TIMEOUT = 'MIDSCENE_MODEL_TIMEOUT';
const MIDSCENE_MODEL_TEMPERATURE = 'MIDSCENE_MODEL_TEMPERATURE';
const MIDSCENE_MODEL_RETRY_COUNT = 'MIDSCENE_MODEL_RETRY_COUNT';
const MIDSCENE_MODEL_RETRY_INTERVAL = 'MIDSCENE_MODEL_RETRY_INTERVAL';
const MIDSCENE_MODEL_REASONING_EFFORT = 'MIDSCENE_MODEL_REASONING_EFFORT';
const MIDSCENE_MODEL_REASONING_ENABLED = 'MIDSCENE_MODEL_REASONING_ENABLED';
const MIDSCENE_MODEL_REASONING_BUDGET = 'MIDSCENE_MODEL_REASONING_BUDGET';
const OPENAI_API_KEY = 'OPENAI_API_KEY';
const OPENAI_BASE_URL = 'OPENAI_BASE_URL';
const MIDSCENE_OPENAI_INIT_CONFIG_JSON = 'MIDSCENE_OPENAI_INIT_CONFIG_JSON';
const MIDSCENE_OPENAI_HTTP_PROXY = 'MIDSCENE_OPENAI_HTTP_PROXY';
const MIDSCENE_OPENAI_SOCKS_PROXY = 'MIDSCENE_OPENAI_SOCKS_PROXY';
const OPENAI_MAX_TOKENS = 'OPENAI_MAX_TOKENS';
const MIDSCENE_ADB_PATH = 'MIDSCENE_ADB_PATH';
const MIDSCENE_ADB_REMOTE_HOST = 'MIDSCENE_ADB_REMOTE_HOST';
const MIDSCENE_ADB_REMOTE_PORT = 'MIDSCENE_ADB_REMOTE_PORT';
const MIDSCENE_ANDROID_IME_STRATEGY = 'MIDSCENE_ANDROID_IME_STRATEGY';
const MIDSCENE_IOS_DEVICE_UDID = 'MIDSCENE_IOS_DEVICE_UDID';
const MIDSCENE_IOS_SIMULATOR_UDID = 'MIDSCENE_IOS_SIMULATOR_UDID';
const MIDSCENE_CACHE = 'MIDSCENE_CACHE';
const MIDSCENE_USE_VLM_UI_TARS = 'MIDSCENE_USE_VLM_UI_TARS';
const MIDSCENE_USE_QWEN_VL = 'MIDSCENE_USE_QWEN_VL';
const MIDSCENE_USE_QWEN3_VL = 'MIDSCENE_USE_QWEN3_VL';
const MIDSCENE_USE_DOUBAO_VISION = 'MIDSCENE_USE_DOUBAO_VISION';
const MIDSCENE_USE_GEMINI = 'MIDSCENE_USE_GEMINI';
const MIDSCENE_USE_VL_MODEL = 'MIDSCENE_USE_VL_MODEL';
const MATCH_BY_POSITION = 'MATCH_BY_POSITION';
const MIDSCENE_REPORT_TAG_NAME = 'MIDSCENE_REPORT_TAG_NAME';
const MIDSCENE_REPORT_QUIET = 'MIDSCENE_REPORT_QUIET';
const MIDSCENE_PREFERRED_LANGUAGE = 'MIDSCENE_PREFERRED_LANGUAGE';
const MIDSCENE_CACHE_MAX_FILENAME_LENGTH = 'MIDSCENE_CACHE_MAX_FILENAME_LENGTH';
const MIDSCENE_REPLANNING_CYCLE_LIMIT = 'MIDSCENE_REPLANNING_CYCLE_LIMIT';
const MIDSCENE_RUN_DIR = 'MIDSCENE_RUN_DIR';
const MIDSCENE_INSIGHT_MODEL_NAME = 'MIDSCENE_INSIGHT_MODEL_NAME';
const MIDSCENE_INSIGHT_MODEL_SOCKS_PROXY = 'MIDSCENE_INSIGHT_MODEL_SOCKS_PROXY';
const MIDSCENE_INSIGHT_MODEL_HTTP_PROXY = 'MIDSCENE_INSIGHT_MODEL_HTTP_PROXY';
const MIDSCENE_INSIGHT_MODEL_BASE_URL = 'MIDSCENE_INSIGHT_MODEL_BASE_URL';
const MIDSCENE_INSIGHT_MODEL_API_KEY = 'MIDSCENE_INSIGHT_MODEL_API_KEY';
const MIDSCENE_INSIGHT_MODEL_INIT_CONFIG_JSON = 'MIDSCENE_INSIGHT_MODEL_INIT_CONFIG_JSON';
const MIDSCENE_INSIGHT_MODEL_EXTRA_BODY_JSON = 'MIDSCENE_INSIGHT_MODEL_EXTRA_BODY_JSON';
const MIDSCENE_INSIGHT_MODEL_TIMEOUT = 'MIDSCENE_INSIGHT_MODEL_TIMEOUT';
const MIDSCENE_INSIGHT_MODEL_TEMPERATURE = 'MIDSCENE_INSIGHT_MODEL_TEMPERATURE';
const MIDSCENE_INSIGHT_MODEL_RETRY_COUNT = 'MIDSCENE_INSIGHT_MODEL_RETRY_COUNT';
const MIDSCENE_INSIGHT_MODEL_RETRY_INTERVAL = 'MIDSCENE_INSIGHT_MODEL_RETRY_INTERVAL';
const MIDSCENE_INSIGHT_MODEL_FAMILY = 'MIDSCENE_INSIGHT_MODEL_FAMILY';
const MIDSCENE_INSIGHT_MODEL_REASONING_EFFORT = 'MIDSCENE_INSIGHT_MODEL_REASONING_EFFORT';
const MIDSCENE_INSIGHT_MODEL_REASONING_ENABLED = 'MIDSCENE_INSIGHT_MODEL_REASONING_ENABLED';
const MIDSCENE_INSIGHT_MODEL_REASONING_BUDGET = 'MIDSCENE_INSIGHT_MODEL_REASONING_BUDGET';
const MIDSCENE_PLANNING_MODEL_NAME = 'MIDSCENE_PLANNING_MODEL_NAME';
const MIDSCENE_PLANNING_MODEL_SOCKS_PROXY = 'MIDSCENE_PLANNING_MODEL_SOCKS_PROXY';
const MIDSCENE_PLANNING_MODEL_HTTP_PROXY = 'MIDSCENE_PLANNING_MODEL_HTTP_PROXY';
const MIDSCENE_PLANNING_MODEL_BASE_URL = 'MIDSCENE_PLANNING_MODEL_BASE_URL';
const MIDSCENE_PLANNING_MODEL_API_KEY = 'MIDSCENE_PLANNING_MODEL_API_KEY';
const MIDSCENE_PLANNING_MODEL_INIT_CONFIG_JSON = 'MIDSCENE_PLANNING_MODEL_INIT_CONFIG_JSON';
const MIDSCENE_PLANNING_MODEL_EXTRA_BODY_JSON = 'MIDSCENE_PLANNING_MODEL_EXTRA_BODY_JSON';
const MIDSCENE_PLANNING_MODEL_TIMEOUT = 'MIDSCENE_PLANNING_MODEL_TIMEOUT';
const MIDSCENE_PLANNING_MODEL_TEMPERATURE = 'MIDSCENE_PLANNING_MODEL_TEMPERATURE';
const MIDSCENE_PLANNING_MODEL_RETRY_COUNT = 'MIDSCENE_PLANNING_MODEL_RETRY_COUNT';
const MIDSCENE_PLANNING_MODEL_RETRY_INTERVAL = 'MIDSCENE_PLANNING_MODEL_RETRY_INTERVAL';
const MIDSCENE_PLANNING_MODEL_FAMILY = 'MIDSCENE_PLANNING_MODEL_FAMILY';
const MIDSCENE_PLANNING_MODEL_REASONING_EFFORT = 'MIDSCENE_PLANNING_MODEL_REASONING_EFFORT';
const MIDSCENE_PLANNING_MODEL_REASONING_ENABLED = 'MIDSCENE_PLANNING_MODEL_REASONING_ENABLED';
const MIDSCENE_PLANNING_MODEL_REASONING_BUDGET = 'MIDSCENE_PLANNING_MODEL_REASONING_BUDGET';
const MIDSCENE_MODEL_FAMILY = 'MIDSCENE_MODEL_FAMILY';
const UNUSED_ENV_KEYS = [
    MIDSCENE_DANGEROUSLY_PRINT_ALL_CONFIG
];
const BASIC_ENV_KEYS = [
    MIDSCENE_DEBUG_MODE,
    MIDSCENE_DEBUG_MODEL_PROFILE,
    MIDSCENE_DEBUG_MODEL_RESPONSE,
    MIDSCENE_RUN_DIR
];
const BOOLEAN_ENV_KEYS = [
    MIDSCENE_CACHE,
    MIDSCENE_MCP_USE_PUPPETEER_MODE,
    MIDSCENE_MCP_ANDROID_MODE,
    MIDSCENE_LANGSMITH_DEBUG,
    MIDSCENE_LANGFUSE_DEBUG,
    MIDSCENE_REPORT_QUIET
];
const NUMBER_ENV_KEYS = [
    MIDSCENE_MODEL_MAX_TOKENS,
    MIDSCENE_CACHE_MAX_FILENAME_LENGTH,
    MIDSCENE_REPLANNING_CYCLE_LIMIT
];
const STRING_ENV_KEYS = [
    MIDSCENE_MODEL_MAX_TOKENS,
    OPENAI_MAX_TOKENS,
    MIDSCENE_ADB_PATH,
    MIDSCENE_ADB_REMOTE_HOST,
    MIDSCENE_ADB_REMOTE_PORT,
    MIDSCENE_ANDROID_IME_STRATEGY,
    MIDSCENE_IOS_DEVICE_UDID,
    MIDSCENE_IOS_SIMULATOR_UDID,
    MIDSCENE_REPORT_TAG_NAME,
    MIDSCENE_PREFERRED_LANGUAGE,
    MATCH_BY_POSITION,
    MIDSCENE_MCP_CHROME_PATH,
    DOCKER_CONTAINER
];
const GLOBAL_ENV_KEYS = [
    ...BOOLEAN_ENV_KEYS,
    ...NUMBER_ENV_KEYS,
    ...STRING_ENV_KEYS
];
const MODEL_ENV_KEYS = [
    MIDSCENE_MODEL_NAME,
    MIDSCENE_MODEL_INIT_CONFIG_JSON,
    MIDSCENE_MODEL_EXTRA_BODY_JSON,
    MIDSCENE_MODEL_API_KEY,
    MIDSCENE_MODEL_BASE_URL,
    MIDSCENE_MODEL_SOCKS_PROXY,
    MIDSCENE_MODEL_HTTP_PROXY,
    MIDSCENE_MODEL_TIMEOUT,
    MIDSCENE_MODEL_TEMPERATURE,
    MIDSCENE_MODEL_RETRY_COUNT,
    MIDSCENE_MODEL_RETRY_INTERVAL,
    MIDSCENE_MODEL_REASONING_EFFORT,
    MIDSCENE_MODEL_REASONING_ENABLED,
    MIDSCENE_MODEL_REASONING_BUDGET,
    MIDSCENE_USE_VLM_UI_TARS,
    MIDSCENE_USE_QWEN_VL,
    MIDSCENE_USE_QWEN3_VL,
    MIDSCENE_USE_DOUBAO_VISION,
    MIDSCENE_USE_GEMINI,
    MIDSCENE_USE_VL_MODEL,
    OPENAI_API_KEY,
    OPENAI_BASE_URL,
    MIDSCENE_OPENAI_INIT_CONFIG_JSON,
    MIDSCENE_OPENAI_HTTP_PROXY,
    MIDSCENE_OPENAI_SOCKS_PROXY,
    MIDSCENE_INSIGHT_MODEL_NAME,
    MIDSCENE_INSIGHT_MODEL_SOCKS_PROXY,
    MIDSCENE_INSIGHT_MODEL_HTTP_PROXY,
    MIDSCENE_INSIGHT_MODEL_BASE_URL,
    MIDSCENE_INSIGHT_MODEL_API_KEY,
    MIDSCENE_INSIGHT_MODEL_INIT_CONFIG_JSON,
    MIDSCENE_INSIGHT_MODEL_EXTRA_BODY_JSON,
    MIDSCENE_INSIGHT_MODEL_TIMEOUT,
    MIDSCENE_INSIGHT_MODEL_TEMPERATURE,
    MIDSCENE_INSIGHT_MODEL_RETRY_COUNT,
    MIDSCENE_INSIGHT_MODEL_RETRY_INTERVAL,
    MIDSCENE_INSIGHT_MODEL_FAMILY,
    MIDSCENE_INSIGHT_MODEL_REASONING_EFFORT,
    MIDSCENE_INSIGHT_MODEL_REASONING_ENABLED,
    MIDSCENE_INSIGHT_MODEL_REASONING_BUDGET,
    MIDSCENE_PLANNING_MODEL_NAME,
    MIDSCENE_PLANNING_MODEL_SOCKS_PROXY,
    MIDSCENE_PLANNING_MODEL_HTTP_PROXY,
    MIDSCENE_PLANNING_MODEL_BASE_URL,
    MIDSCENE_PLANNING_MODEL_API_KEY,
    MIDSCENE_PLANNING_MODEL_INIT_CONFIG_JSON,
    MIDSCENE_PLANNING_MODEL_EXTRA_BODY_JSON,
    MIDSCENE_PLANNING_MODEL_TIMEOUT,
    MIDSCENE_PLANNING_MODEL_TEMPERATURE,
    MIDSCENE_PLANNING_MODEL_RETRY_COUNT,
    MIDSCENE_PLANNING_MODEL_RETRY_INTERVAL,
    MIDSCENE_PLANNING_MODEL_FAMILY,
    MIDSCENE_PLANNING_MODEL_REASONING_EFFORT,
    MIDSCENE_PLANNING_MODEL_REASONING_ENABLED,
    MIDSCENE_PLANNING_MODEL_REASONING_BUDGET,
    MIDSCENE_MODEL_FAMILY
];
const ALL_ENV_KEYS = [
    ...UNUSED_ENV_KEYS,
    ...BASIC_ENV_KEYS,
    ...GLOBAL_ENV_KEYS,
    ...MODEL_ENV_KEYS
];
const MODEL_FAMILY_VALUES = [
    'doubao-vision',
    'doubao-seed',
    'gemini',
    'qwen2.5-vl',
    'qwen3-vl',
    'qwen3.5',
    'qwen3.6',
    'vlm-ui-tars',
    'vlm-ui-tars-doubao',
    'vlm-ui-tars-doubao-1.5',
    'glm-v',
    'auto-glm',
    'auto-glm-multilingual',
    'gpt-5'
];
var types_UITarsModelVersion = /*#__PURE__*/ function(UITarsModelVersion) {
    UITarsModelVersion["V1_0"] = "1.0";
    UITarsModelVersion["V1_5"] = "1.5";
    UITarsModelVersion["DOUBAO_1_5_15B"] = "doubao-1.5-15B";
    UITarsModelVersion["DOUBAO_1_5_20B"] = "doubao-1.5-20B";
    return UITarsModelVersion;
}({});
export { ALL_ENV_KEYS, BASIC_ENV_KEYS, BOOLEAN_ENV_KEYS, DOCKER_CONTAINER, GLOBAL_ENV_KEYS, MATCH_BY_POSITION, MIDSCENE_ADB_PATH, MIDSCENE_ADB_REMOTE_HOST, MIDSCENE_ADB_REMOTE_PORT, MIDSCENE_ANDROID_IME_STRATEGY, MIDSCENE_CACHE, MIDSCENE_CACHE_MAX_FILENAME_LENGTH, MIDSCENE_DANGEROUSLY_PRINT_ALL_CONFIG, MIDSCENE_DEBUG_MODE, MIDSCENE_DEBUG_MODEL_PROFILE, MIDSCENE_DEBUG_MODEL_RESPONSE, MIDSCENE_INSIGHT_MODEL_API_KEY, MIDSCENE_INSIGHT_MODEL_BASE_URL, MIDSCENE_INSIGHT_MODEL_EXTRA_BODY_JSON, MIDSCENE_INSIGHT_MODEL_FAMILY, MIDSCENE_INSIGHT_MODEL_HTTP_PROXY, MIDSCENE_INSIGHT_MODEL_INIT_CONFIG_JSON, MIDSCENE_INSIGHT_MODEL_NAME, MIDSCENE_INSIGHT_MODEL_REASONING_BUDGET, MIDSCENE_INSIGHT_MODEL_REASONING_EFFORT, MIDSCENE_INSIGHT_MODEL_REASONING_ENABLED, MIDSCENE_INSIGHT_MODEL_RETRY_COUNT, MIDSCENE_INSIGHT_MODEL_RETRY_INTERVAL, MIDSCENE_INSIGHT_MODEL_SOCKS_PROXY, MIDSCENE_INSIGHT_MODEL_TEMPERATURE, MIDSCENE_INSIGHT_MODEL_TIMEOUT, MIDSCENE_IOS_DEVICE_UDID, MIDSCENE_IOS_SIMULATOR_UDID, MIDSCENE_LANGFUSE_DEBUG, MIDSCENE_LANGSMITH_DEBUG, MIDSCENE_MCP_ANDROID_MODE, MIDSCENE_MCP_CHROME_PATH, MIDSCENE_MCP_USE_PUPPETEER_MODE, MIDSCENE_MODEL_API_KEY, MIDSCENE_MODEL_BASE_URL, MIDSCENE_MODEL_EXTRA_BODY_JSON, MIDSCENE_MODEL_FAMILY, MIDSCENE_MODEL_HTTP_PROXY, MIDSCENE_MODEL_INIT_CONFIG_JSON, MIDSCENE_MODEL_MAX_TOKENS, MIDSCENE_MODEL_NAME, MIDSCENE_MODEL_REASONING_BUDGET, MIDSCENE_MODEL_REASONING_EFFORT, MIDSCENE_MODEL_REASONING_ENABLED, MIDSCENE_MODEL_RETRY_COUNT, MIDSCENE_MODEL_RETRY_INTERVAL, MIDSCENE_MODEL_SOCKS_PROXY, MIDSCENE_MODEL_TEMPERATURE, MIDSCENE_MODEL_TIMEOUT, MIDSCENE_OPENAI_HTTP_PROXY, MIDSCENE_OPENAI_INIT_CONFIG_JSON, MIDSCENE_OPENAI_SOCKS_PROXY, MIDSCENE_PLANNING_MODEL_API_KEY, MIDSCENE_PLANNING_MODEL_BASE_URL, MIDSCENE_PLANNING_MODEL_EXTRA_BODY_JSON, MIDSCENE_PLANNING_MODEL_FAMILY, MIDSCENE_PLANNING_MODEL_HTTP_PROXY, MIDSCENE_PLANNING_MODEL_INIT_CONFIG_JSON, MIDSCENE_PLANNING_MODEL_NAME, MIDSCENE_PLANNING_MODEL_REASONING_BUDGET, MIDSCENE_PLANNING_MODEL_REASONING_EFFORT, MIDSCENE_PLANNING_MODEL_REASONING_ENABLED, MIDSCENE_PLANNING_MODEL_RETRY_COUNT, MIDSCENE_PLANNING_MODEL_RETRY_INTERVAL, MIDSCENE_PLANNING_MODEL_SOCKS_PROXY, MIDSCENE_PLANNING_MODEL_TEMPERATURE, MIDSCENE_PLANNING_MODEL_TIMEOUT, MIDSCENE_PREFERRED_LANGUAGE, MIDSCENE_REPLANNING_CYCLE_LIMIT, MIDSCENE_REPORT_QUIET, MIDSCENE_REPORT_TAG_NAME, MIDSCENE_RUN_DIR, MIDSCENE_USE_DOUBAO_VISION, MIDSCENE_USE_GEMINI, MIDSCENE_USE_QWEN3_VL, MIDSCENE_USE_QWEN_VL, MIDSCENE_USE_VLM_UI_TARS, MIDSCENE_USE_VL_MODEL, MODEL_ENV_KEYS, MODEL_FAMILY_VALUES, NUMBER_ENV_KEYS, OPENAI_API_KEY, OPENAI_BASE_URL, OPENAI_MAX_TOKENS, STRING_ENV_KEYS, types_UITarsModelVersion as UITarsModelVersion, UNUSED_ENV_KEYS };
