// @midscene/web@1.7.5/dist/es/mcp-tools-cdp.mjs
import { ScreenshotItem, z } from "@midscene/core";
import { getDebug } from "@midscene/shared/logger";
import { BaseMidsceneTools } from "@midscene/shared/mcp/base-tools";
import puppeteer_core from "puppeteer-core";
import { getProxyEndpoint } from "./cdp-proxy-manager.mjs";
import { cleanupTargetIdFile, readSavedTargetId, saveTargetId } from "./cdp-target-store.mjs";
import { PuppeteerAgent } from "./puppeteer/index.mjs";
import { StaticPage } from "./static/index.mjs";
function _define_property(obj, key, value) {
    if (key in obj) Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
    });
    else obj[key] = value;
    return obj;
}
const debug = getDebug('mcp:cdp');
const CDP_TARGET_DISCOVERY_DELAY_MS = 500;
function getTargetId(page) {
    return page.target()._targetId;
}
class WebCdpMidsceneTools extends BaseMidsceneTools {
    createTemporaryDevice() {
        return new StaticPage({
            screenshot: ScreenshotItem.create('', Date.now()),
            shotSize: {
                width: 1920,
                height: 1080
            },
            shrunkShotToLogicalRatio: 1
        });
    }
    async ensureAgent(navigateToUrl) {
        if (this.agent && navigateToUrl) {
            try {
                await this.agent?.destroy?.();
            } catch (error) {
                console.debug('Failed to destroy agent during re-init:', error);
            }
            this.agent = void 0;
        }
        if (this.agent) return this.agent;
        if (!this.activeBrowser) {
            const endpoint = await getProxyEndpoint(this.cdpEndpoint);
            this.activeBrowser = await puppeteer_core.connect({
                browserWSEndpoint: endpoint,
                defaultViewport: null
            });
        }
        const browser = this.activeBrowser;
        let pages = await browser.pages();
        if (0 === pages.length) {
            await new Promise((r)=>setTimeout(r, CDP_TARGET_DISCOVERY_DELAY_MS));
            pages = await browser.pages();
        }
        const webPages = pages.filter((p)=>/^https?:\/\//.test(p.url()));
        debug('Found %d page(s), %d web page(s): %o', pages.length, webPages.length, pages.map((p)=>p.url()));
        let page;
        if (navigateToUrl) if (webPages.length > 0) {
            page = webPages[webPages.length - 1];
            await page.bringToFront();
            await page.goto(navigateToUrl, {
                timeout: 30000,
                waitUntil: 'domcontentloaded'
            });
        } else {
            page = await browser.newPage();
            await page.goto(navigateToUrl, {
                timeout: 30000,
                waitUntil: 'domcontentloaded'
            });
        }
        else {
            const savedTargetId = readSavedTargetId();
            let matchedPage;
            if (savedTargetId && pages.length > 0) {
                matchedPage = pages.find((p)=>getTargetId(p) === savedTargetId);
                matchedPage ? debug('Matched saved targetId %s', savedTargetId) : debug('Saved targetId %s not found among %d pages, falling back', savedTargetId, pages.length);
            }
            page = matchedPage ? matchedPage : webPages.length > 0 ? webPages[webPages.length - 1] : pages.length > 0 ? pages[pages.length - 1] : await browser.newPage();
            await page.bringToFront();
        }
        const targetId = getTargetId(page);
        if (targetId) saveTargetId(targetId);
        else debug('No targetId on page.target(); cross-command tab reuse disabled until puppeteer integration is updated.');
        this.agent = new PuppeteerAgent(page);
        return this.agent;
    }
    async destroy() {
        await super.destroy();
        if (this.activeBrowser) {
            this.activeBrowser.disconnect();
            this.activeBrowser = null;
        }
    }
    preparePlatformTools() {
        return [
            {
                name: 'web_connect',
                description: 'Connect to a web page via CDP. Opens a new tab with the given URL, or reuses the current page.',
                schema: {
                    url: z.string().url().optional().describe('URL to open in new tab (omit to use current page)')
                },
                handler: async (args)=>{
                    const { url } = args;
                    if (this.agent) {
                        try {
                            await this.agent.destroy?.();
                        } catch (e) {
                            console.debug('Failed to destroy agent during connect:', e);
                        }
                        this.agent = void 0;
                    }
                    this.agent = await this.ensureAgent(url);
                    const screenshot = await this.agent.page?.screenshotBase64();
                    const label = url ?? 'current page';
                    return {
                        content: [
                            {
                                type: 'text',
                                text: `Connected via CDP to: ${label}`
                            },
                            ...screenshot ? this.buildScreenshotContent(screenshot) : []
                        ]
                    };
                }
            },
            {
                name: 'web_disconnect',
                description: 'Disconnect from current web page. The browser stays running (managed externally).',
                schema: {},
                handler: async ()=>{
                    if (this.agent) {
                        try {
                            await this.agent.destroy?.();
                        } catch (e) {
                            console.debug('Failed to destroy agent during disconnect:', e);
                        }
                        this.agent = void 0;
                    }
                    if (this.activeBrowser) {
                        this.activeBrowser.disconnect();
                        this.activeBrowser = null;
                    }
                    cleanupTargetIdFile();
                    return this.buildTextResult('Disconnected from web page (browser still running externally)');
                }
            }
        ];
    }
    constructor(cdpEndpoint){
        super(), _define_property(this, "cdpEndpoint", void 0), _define_property(this, "activeBrowser", null);
        this.cdpEndpoint = cdpEndpoint;
    }
}
export { WebCdpMidsceneTools };
