// @midscene/web@1.7.5/dist/es/playwright/ai-fixture.mjs
import { PlaywrightAgent } from "./index.mjs";
import { processCacheConfig } from "@midscene/core/utils";
import { DEFAULT_WAIT_FOR_NAVIGATION_TIMEOUT, DEFAULT_WAIT_FOR_NETWORK_IDLE_TIMEOUT } from "@midscene/shared/constants";
import { getDebug } from "@midscene/shared/logger";
import { replaceIllegalPathCharsAndSpace, uuid } from "@midscene/shared/utils";
import { test } from "@playwright/test";
const debugPage = getDebug('web:playwright:ai-fixture');
const groupAndCaseForTest = (testInfo)=>{
    let taskFile;
    let taskTitle;
    const titlePath = [
        ...testInfo.titlePath
    ];
    if (titlePath.length > 1) {
        taskFile = titlePath.shift() || 'unnamed';
        taskTitle = titlePath.join('__');
    } else if (1 === titlePath.length) {
        taskTitle = titlePath[0];
        taskFile = `${taskTitle}`;
    } else {
        taskTitle = 'unnamed';
        taskFile = 'unnamed';
    }
    const taskTitleWithRetry = `${taskTitle}${testInfo.retry ? `(retry #${testInfo.retry})` : ''}`;
    return {
        file: taskFile,
        id: replaceIllegalPathCharsAndSpace(`${taskFile}(${taskTitle})`),
        title: replaceIllegalPathCharsAndSpace(taskTitleWithRetry)
    };
};
const midsceneAgentKeyId = '_midsceneAgentId';
const midsceneDumpAnnotationId = 'MIDSCENE_DUMP_ANNOTATION';
const PlaywrightAiFixture = (options)=>{
    const { forceSameTabNavigation = true, waitForNetworkIdleTimeout = DEFAULT_WAIT_FOR_NETWORK_IDLE_TIMEOUT, waitForNavigationTimeout = DEFAULT_WAIT_FOR_NAVIGATION_TIMEOUT, cache, ...sharedAgentOptions } = options ?? {};
    const processTestCacheConfig = (testInfo)=>{
        const { id } = groupAndCaseForTest(testInfo);
        return processCacheConfig(cache, id);
    };
    const pageAgentMap = {};
    const testAgentRecords = new Map();
    const getAgentRecordsForTest = (testInfo)=>{
        let records = testAgentRecords.get(testInfo.testId);
        if (!records) {
            records = new Map();
            testAgentRecords.set(testInfo.testId, records);
        }
        return records;
    };
    const setReportAnnotation = (testInfo, reportPaths)=>{
        testInfo.annotations = testInfo.annotations.filter((item)=>item.type !== midsceneDumpAnnotationId);
        for (const reportPath of reportPaths)testInfo.annotations.push({
            type: midsceneDumpAnnotationId,
            description: reportPath
        });
    };
    const finalizeAgentRecord = async (record)=>{
        if (!record.finalizePromise) record.finalizePromise = (async ()=>{
            await record.agent.destroy();
            const reportPath = record.agent.reportFile || void 0;
            record.finalReportPath = reportPath;
            return reportPath;
        })();
        return await record.finalizePromise;
    };
    const createOrReuseAgentForPage = (page, testInfo, opts)=>{
        let idForPage = page[midsceneAgentKeyId];
        if (!idForPage) {
            idForPage = uuid();
            page[midsceneAgentKeyId] = idForPage;
            const { file, title } = groupAndCaseForTest(testInfo);
            const cacheConfig = processTestCacheConfig(testInfo);
            const reportTag = `playwright-${title.replace(/[\\/]/g, '-')}-${idForPage}`;
            const agent = new PlaywrightAgent(page, {
                testId: reportTag,
                reportFileName: reportTag,
                forceSameTabNavigation,
                cache: cacheConfig,
                groupName: title,
                groupDescription: file,
                generateReport: true,
                ...sharedAgentOptions,
                ...opts
            });
            pageAgentMap[idForPage] = agent;
            const records = getAgentRecordsForTest(testInfo);
            const record = {
                agent
            };
            records.set(idForPage, record);
            page.on('close', async ()=>{
                debugPage('page closed');
                try {
                    await finalizeAgentRecord(record);
                } finally{
                    delete pageAgentMap[idForPage];
                }
            });
        }
        return pageAgentMap[idForPage];
    };
    async function generateAiFunction(options) {
        const { page, testInfo, use, aiActionType } = options;
        const agent = createOrReuseAgentForPage(page, testInfo, {
            waitForNavigationTimeout,
            waitForNetworkIdleTimeout
        });
        await use(async (taskPrompt, ...args)=>new Promise((resolve, reject)=>{
                test.step(`ai-${aiActionType} - ${JSON.stringify(taskPrompt)}`, async ()=>{
                    try {
                        debugPage(`waitForNetworkIdle timeout: ${waitForNetworkIdleTimeout}`);
                        await agent.waitForNetworkIdle(waitForNetworkIdleTimeout);
                    } catch (error) {
                        console.warn('[midscene:warning] Waiting for network idle has timed out, but Midscene will continue execution. Please check https://midscenejs.com/faq.html#customize-the-network-timeout for more information on customizing the network timeout');
                    }
                    try {
                        const result = await agent[aiActionType].bind(agent)(taskPrompt, ...args);
                        resolve(result);
                    } catch (error) {
                        reject(error);
                    }
                });
            }));
    }
    return {
        _midsceneFinalizeReports: [
            async ({}, use, testInfo)=>{
                await use();
                const records = testAgentRecords.get(testInfo.testId);
                if (!records || 0 === records.size) return;
                const reportPaths = (await Promise.all(Array.from(records.values()).map((record)=>finalizeAgentRecord(record)))).filter((reportPath)=>Boolean(reportPath));
                if (reportPaths.length > 0) setReportAnnotation(testInfo, reportPaths);
                testAgentRecords.delete(testInfo.testId);
            },
            {
                auto: true
            }
        ],
        agentForPage: async ({ page }, use, testInfo)=>{
            await use(async (propsPage, opts)=>{
                const cacheConfig = processTestCacheConfig(testInfo);
                let finalCacheConfig = cacheConfig;
                if (opts?.cache !== void 0) {
                    const userCache = opts.cache;
                    if (false === userCache) finalCacheConfig = false;
                    else if (true === userCache) {
                        const { id } = groupAndCaseForTest(testInfo);
                        finalCacheConfig = {
                            id
                        };
                    } else if ('object' == typeof userCache) if (userCache.id) finalCacheConfig = userCache;
                    else {
                        const { id } = groupAndCaseForTest(testInfo);
                        finalCacheConfig = {
                            ...userCache,
                            id
                        };
                    }
                }
                const agent = createOrReuseAgentForPage(propsPage || page, testInfo, {
                    waitForNavigationTimeout,
                    waitForNetworkIdleTimeout,
                    cache: finalCacheConfig,
                    ...opts
                });
                return agent;
            });
        },
        ai: async ({ page }, use, testInfo)=>{
            await generateAiFunction({
                page,
                testInfo,
                use,
                aiActionType: 'ai'
            });
        },
        aiAct: async ({ page }, use, testInfo)=>{
            await generateAiFunction({
                page,
                testInfo,
                use,
                aiActionType: 'aiAct'
            });
        },
        aiAction: async ({ page }, use, testInfo)=>{
            await generateAiFunction({
                page,
                testInfo,
                use,
                aiActionType: 'aiAction'
            });
        },
        aiTap: async ({ page }, use, testInfo)=>{
            await generateAiFunction({
                page,
                testInfo,
                use,
                aiActionType: 'aiTap'
            });
        },
        aiRightClick: async ({ page }, use, testInfo)=>{
            await generateAiFunction({
                page,
                testInfo,
                use,
                aiActionType: 'aiRightClick'
            });
        },
        aiDoubleClick: async ({ page }, use, testInfo)=>{
            await generateAiFunction({
                page,
                testInfo,
                use,
                aiActionType: 'aiDoubleClick'
            });
        },
        aiHover: async ({ page }, use, testInfo)=>{
            await generateAiFunction({
                page,
                testInfo,
                use,
                aiActionType: 'aiHover'
            });
        },
        aiInput: async ({ page }, use, testInfo)=>{
            await generateAiFunction({
                page,
                testInfo,
                use,
                aiActionType: 'aiInput'
            });
        },
        aiKeyboardPress: async ({ page }, use, testInfo)=>{
            await generateAiFunction({
                page,
                testInfo,
                use,
                aiActionType: 'aiKeyboardPress'
            });
        },
        aiScroll: async ({ page }, use, testInfo)=>{
            await generateAiFunction({
                page,
                testInfo,
                use,
                aiActionType: 'aiScroll'
            });
        },
        aiQuery: async ({ page }, use, testInfo)=>{
            await generateAiFunction({
                page,
                testInfo,
                use,
                aiActionType: 'aiQuery'
            });
        },
        aiAssert: async ({ page }, use, testInfo)=>{
            await generateAiFunction({
                page,
                testInfo,
                use,
                aiActionType: 'aiAssert'
            });
        },
        aiWaitFor: async ({ page }, use, testInfo)=>{
            await generateAiFunction({
                page,
                testInfo,
                use,
                aiActionType: 'aiWaitFor'
            });
        },
        aiLocate: async ({ page }, use, testInfo)=>{
            await generateAiFunction({
                page,
                testInfo,
                use,
                aiActionType: 'aiLocate'
            });
        },
        aiNumber: async ({ page }, use, testInfo)=>{
            await generateAiFunction({
                page,
                testInfo,
                use,
                aiActionType: 'aiNumber'
            });
        },
        aiString: async ({ page }, use, testInfo)=>{
            await generateAiFunction({
                page,
                testInfo,
                use,
                aiActionType: 'aiString'
            });
        },
        aiBoolean: async ({ page }, use, testInfo)=>{
            await generateAiFunction({
                page,
                testInfo,
                use,
                aiActionType: 'aiBoolean'
            });
        },
        aiAsk: async ({ page }, use, testInfo)=>{
            await generateAiFunction({
                page,
                testInfo,
                use,
                aiActionType: 'aiAsk'
            });
        },
        runYaml: async ({ page }, use, testInfo)=>{
            await generateAiFunction({
                page,
                testInfo,
                use,
                aiActionType: 'runYaml'
            });
        },
        setAIActionContext: async ({ page }, use, testInfo)=>{
            await generateAiFunction({
                page,
                testInfo,
                use,
                aiActionType: 'setAIActionContext'
            });
        },
        evaluateJavaScript: async ({ page }, use, testInfo)=>{
            await generateAiFunction({
                page,
                testInfo,
                use,
                aiActionType: 'evaluateJavaScript'
            });
        },
        recordToReport: async ({ page }, use, testInfo)=>{
            await generateAiFunction({
                page,
                testInfo,
                use,
                aiActionType: 'recordToReport'
            });
        },
        logScreenshot: async ({ page }, use, testInfo)=>{
            await generateAiFunction({
                page,
                testInfo,
                use,
                aiActionType: 'logScreenshot'
            });
        },
        freezePageContext: async ({ page }, use, testInfo)=>{
            await generateAiFunction({
                page,
                testInfo,
                use,
                aiActionType: 'freezePageContext'
            });
        },
        unfreezePageContext: async ({ page }, use, testInfo)=>{
            await generateAiFunction({
                page,
                testInfo,
                use,
                aiActionType: 'unfreezePageContext'
            });
        }
    };
};
export { PlaywrightAiFixture, midsceneDumpAnnotationId };
