// @midscene/web@1.7.5/dist/es/playwright/index.mjs
import { Agent } from "@midscene/core/agent";
import { WebPage } from "./page.mjs";
import { PlaywrightAiFixture } from "./ai-fixture.mjs";
import { overrideAIConfig } from "@midscene/shared/env";
import { getDebug } from "@midscene/shared/logger";
import semver from "semver";
import { BROWSER_NAVIGATION_ERROR_PATTERN, forceChromeSelectRendering as base_page_mjs_forceChromeSelectRendering, forceClosePopup } from "../puppeteer/base-page.mjs";
import { getWebpackRequire } from "../utils.mjs";
const debug = getDebug('playwright:agent');
function getPlaywrightVersion() {
    try {
        const playwrightPkg = getWebpackRequire()('playwright/package.json');
        return playwrightPkg.version || null;
    } catch (error) {
        console.error('[midscene:error] Failed to get Playwright version', error);
        return null;
    }
}
class PlaywrightAgent extends Agent {
    isRetryableContextError(error) {
        return error instanceof Error && BROWSER_NAVIGATION_ERROR_PATTERN.test(error.message);
    }
    async waitForNetworkIdle(timeout = 1000) {
        await this.page.underlyingPage.waitForLoadState('networkidle', {
            timeout
        });
    }
    constructor(page, opts){
        if (!page) throw new Error('[midscene] PlaywrightAgent requires a valid Playwright page instance. Please make sure to pass a valid page object.');
        const webPage = new WebPage(page, opts);
        super(webPage, opts);
        const { forceSameTabNavigation = true, forceChromeSelectRendering } = opts ?? {};
        if (forceSameTabNavigation) forceClosePopup(page, debug);
        if (forceChromeSelectRendering) {
            const playwrightVersion = getPlaywrightVersion();
            if (playwrightVersion && !semver.gte(playwrightVersion, '1.52.0')) console.warn(`[midscene:error] forceChromeSelectRendering requires Playwright >= 1.52.0, but current version is ${playwrightVersion}. This feature may not work correctly.`);
            base_page_mjs_forceChromeSelectRendering(page);
        }
    }
}
export { PlaywrightAgent, PlaywrightAiFixture, WebPage as PlaywrightWebPage, overrideAIConfig };
