// @midscene/shared/dist/es/cli/cli-args.mjs
import { z } from "zod";
import { getKeyAliases } from "../key-alias-utils.mjs";
function parseValue(raw) {
    if (raw.startsWith('{') || raw.startsWith('[')) try {
        return JSON.parse(raw);
    } catch  {}
    if (/^-?\d+(\.\d+)?$/.test(raw)) return Number(raw);
    return raw;
}
function walkCliArgs(args, setArgValue) {
    for(let i = 0; i < args.length; i++){
        const arg = args[i];
        if (!arg.startsWith('--')) continue;
        const body = arg.slice(2);
        const eqIdx = body.indexOf('=');
        if (eqIdx >= 0) setArgValue(body.slice(0, eqIdx), parseValue(body.slice(eqIdx + 1)));
        else if (args[i + 1] && !args[i + 1].startsWith('--')) {
            i++;
            setArgValue(body, parseValue(args[i]));
        } else setArgValue(body, true);
    }
}
function parseCliArgs(args) {
    const result = {};
    walkCliArgs(args, (key, value)=>{
        result[key] = value;
    });
    return result;
}
function formatCliOptionName(name) {
    return `--${name}`;
}
function getCliOptionDisplay(key, cliOption) {
    const label = formatCliOptionName(cliOption?.preferredName ?? key);
    const aliases = [
        ...new Set(cliOption?.aliases ?? [])
    ].map((alias)=>formatCliOptionName(alias)).filter((alias)=>alias !== label);
    return {
        label,
        aliases
    };
}
function getAcceptedCliOptionNames(key, cliOption) {
    return [
        ...new Set(cliOption ? [
            cliOption.preferredName ?? key,
            ...cliOption.aliases ?? []
        ] : [
            key,
            ...getKeyAliases(key)
        ])
    ];
}
function toOptionalCliSchemaField(field) {
    if ('object' == typeof field && null !== field && 'function' == typeof field.optional) return field.optional();
    const description = 'object' == typeof field && null !== field && "description" in field && 'string' == typeof field.description ? field.description : void 0;
    return description ? z.any().describe(description) : z.any();
}
function buildCliArgSchema(def) {
    return Object.fromEntries(Object.entries(def.schema).flatMap(([key, zodType])=>getAcceptedCliOptionNames(key, def.cli?.options?.[key]).map((cliKey)=>[
                cliKey,
                toOptionalCliSchemaField(zodType)
            ])));
}
function buildDisallowedCliSpellings(def) {
    const disallowedSpellings = new Set();
    for (const [key] of Object.entries(def.schema)){
        const cliOption = def.cli?.options?.[key];
        const acceptedNames = new Set(getAcceptedCliOptionNames(key, cliOption));
        const knownSpellings = new Set([
            key,
            ...getKeyAliases(key),
            ...cliOption?.preferredName ? getKeyAliases(cliOption.preferredName) : [],
            ...cliOption?.aliases ?? []
        ]);
        for (const spelling of knownSpellings)if (!acceptedNames.has(spelling)) disallowedSpellings.add(spelling);
    }
    return disallowedSpellings;
}
function formatCliValidationError(scriptName, commandName, def, rawArgs) {
    if (0 === Object.keys(def.schema).length) return;
    const cliSchema = z.object(buildCliArgSchema(def)).strict();
    const parsed = cliSchema.safeParse(rawArgs);
    if (parsed.success) return;
    const disallowedSpellings = buildDisallowedCliSpellings(def);
    const unknownKeys = parsed.error.issues.flatMap((issue)=>'unrecognized_keys' === issue.code ? issue.keys : []);
    if (unknownKeys.length > 0) return unknownKeys.map((key)=>{
        if (disallowedSpellings.has(key)) return `Unsupported option "--${key}" for ${scriptName} ${commandName}.`;
        return `Unknown option "--${key}" for ${scriptName} ${commandName}.`;
    }).join('\n');
    const [issue] = parsed.error.issues;
    const optionName = 'string' == typeof issue?.path[0] ? `--${issue.path[0]}` : 'CLI arguments';
    return `Invalid value for "${optionName}" in ${scriptName} ${commandName}: ${issue?.message ?? parsed.error.message}`;
}
export { formatCliValidationError, getCliOptionDisplay, parseCliArgs, parseValue };
