// @midscene/web/dist/es/common/cache-helper.mjs
import { AiJudgeOrderSensitive, callAIWithObjectResponse } from "@midscene/core/ai-model";
const sanitizeXpaths = (xpaths)=>{
    if (!Array.isArray(xpaths)) return [];
    return xpaths.filter((xpath)=>'string' == typeof xpath && xpath.length > 0);
};
async function judgeOrderSensitive(options, debug) {
    if (!options?.targetDescription || !options?.modelConfig) return false;
    try {
        const judgeResult = await AiJudgeOrderSensitive(options.targetDescription, callAIWithObjectResponse, options.modelConfig);
        debug("judged isOrderSensitive=%s for description: %s", judgeResult.isOrderSensitive, options.targetDescription);
        return judgeResult.isOrderSensitive;
    } catch (error) {
        debug('Failed to judge isOrderSensitive: %O', error);
        return false;
    }
}
function buildRectFromElementInfo(elementInfo) {
    const matchedRect = {
        left: elementInfo.rect.left,
        top: elementInfo.rect.top,
        width: elementInfo.rect.width,
        height: elementInfo.rect.height
    };
    return matchedRect;
}
export { buildRectFromElementInfo, judgeOrderSensitive, sanitizeXpaths };
