// @midscene/web@1.7.5/dist/es/bridge-mode/page-browser-side.mjs
import { assert } from "@midscene/shared/utils";
import page from "../chrome-extension/page.mjs";
import { BridgeEvent, DefaultBridgeServerPort, KeyboardEvent, MouseEvent } from "./common.mjs";
import { BridgeClient } from "./io-client.mjs";
function _define_property(obj, key, value) {
    if (key in obj) Object.defineProperty(obj, key, {
        value: value,
        enumerable: true,
        configurable: true,
        writable: true
    });
    else obj[key] = value;
    return obj;
}
class ExtensionBridgePageBrowserSide extends page {
    async setupBridgeClient() {
        const endpoint = this.serverEndpoint || `ws://localhost:${DefaultBridgeServerPort}`;
        let resolveConfirmationGate = ()=>{};
        if (this.onConnectionRequest) this.confirmationPromise = new Promise((resolve)=>{
            resolveConfirmationGate = resolve;
        });
        this.bridgeClient = new BridgeClient(endpoint, async (method, args)=>{
            if (this.confirmationPromise) {
                const allowed = await this.confirmationPromise;
                if (!allowed) throw new Error('Connection denied by user');
            }
            this.onLogMessage(`bridge call from cli side: ${method}`, 'log');
            if (method === BridgeEvent.ConnectNewTabWithUrl) return this.connectNewTabWithUrl.apply(this, args);
            if (method === BridgeEvent.GetBrowserTabList) return this.getBrowserTabList.apply(this, args);
            if (method === BridgeEvent.SetActiveTabId) return this.setActiveTabId.apply(this, args);
            if (method === BridgeEvent.ConnectCurrentTab) return this.connectCurrentTab.apply(this, args);
            if (method === BridgeEvent.UpdateAgentStatus) return this.onLogMessage(args[0], 'status');
            const tabId = await this.getActiveTabId();
            if (!tabId || 0 === tabId) throw new Error('no tab is connected');
            if (method.startsWith(MouseEvent.PREFIX)) {
                const actionName = method.split('.')[1];
                return this.mouse[actionName].apply(this.mouse, args);
            }
            if (method.startsWith(KeyboardEvent.PREFIX)) {
                const actionName = method.split('.')[1];
                return this.keyboard[actionName].apply(this.keyboard, args);
            }
            if (!this[method]) return void this.onLogMessage(`method not found: ${method}`, 'log');
            try {
                const result = await this[method](...args);
                return result;
            } catch (e) {
                const errorMessage = e instanceof Error ? e.message : 'Unknown error';
                this.onLogMessage(`Error calling method: ${method}, ${errorMessage}`, 'log');
                throw new Error(errorMessage, {
                    cause: e
                });
            }
        }, ()=>this.destroy());
        await this.bridgeClient.connect();
        if (this.onConnectionRequest) {
            this.onLogMessage('Waiting for user confirmation...', 'log');
            const allowed = await this.onConnectionRequest();
            resolveConfirmationGate(allowed);
            this.confirmationPromise = null;
            if (!allowed) {
                this.onLogMessage('Connection denied by user', 'log');
                this.bridgeClient.disconnect();
                this.bridgeClient = null;
                throw new Error('Connection denied by user');
            }
        }
        this.onLogMessage(`Bridge connected, cli-side version v${this.bridgeClient.serverVersion}, browser-side version v1.7.5`, 'log');
    }
    async connect() {
        return await this.setupBridgeClient();
    }
    async connectNewTabWithUrl(url, options = {
        forceSameTabNavigation: true
    }) {
        const tab = await chrome.tabs.create({
            url
        });
        const tabId = tab.id;
        assert(tabId, 'failed to get tabId after creating a new tab');
        this.onLogMessage(`Creating new tab: ${url}`, 'log');
        this.newlyCreatedTabIds.push(tabId);
        if (options?.forceSameTabNavigation) this.forceSameTabNavigation = true;
        await this.setActiveTabId(tabId);
    }
    async connectCurrentTab(options = {
        forceSameTabNavigation: true
    }) {
        const tabs = await chrome.tabs.query({
            active: true,
            currentWindow: true
        });
        const tabId = tabs[0]?.id;
        assert(tabId, 'failed to get tabId');
        this.onLogMessage(`Connected to current tab: ${tabs[0]?.url}`, 'log');
        if (options?.forceSameTabNavigation) this.forceSameTabNavigation = true;
        await this.setActiveTabId(tabId);
    }
    async setDestroyOptions(options) {
        this.destroyOptions = options;
    }
    async destroy() {
        if (this.destroyOptions?.closeTab && this.newlyCreatedTabIds.length > 0) {
            this.onLogMessage('Closing all newly created tabs by bridge...', 'log');
            for (const tabId of this.newlyCreatedTabIds)await chrome.tabs.remove(tabId);
            this.newlyCreatedTabIds = [];
        }
        await super.destroy();
        if (this.bridgeClient) {
            this.bridgeClient.disconnect();
            this.bridgeClient = null;
            this.onDisconnect();
        }
    }
    constructor(serverEndpoint, onDisconnect = ()=>{}, onLogMessage = ()=>{}, forceSameTabNavigation = true, onConnectionRequest){
        super(forceSameTabNavigation), _define_property(this, "serverEndpoint", void 0), _define_property(this, "onDisconnect", void 0), _define_property(this, "onLogMessage", void 0), _define_property(this, "onConnectionRequest", void 0), _define_property(this, "bridgeClient", void 0), _define_property(this, "destroyOptions", void 0), _define_property(this, "newlyCreatedTabIds", void 0), _define_property(this, "confirmationPromise", void 0), this.serverEndpoint = serverEndpoint, this.onDisconnect = onDisconnect, this.onLogMessage = onLogMessage, this.onConnectionRequest = onConnectionRequest, this.bridgeClient = null, this.newlyCreatedTabIds = [], this.confirmationPromise = null;
    }
}
export { ExtensionBridgePageBrowserSide };
