// @midscene/web/dist/es/cdp-proxy-manager.mjs
import { spawn } from "node:child_process";
import { existsSync, readFileSync, unlinkSync } from "node:fs";
import node_http from "node:http";
import { join } from "node:path";
import { getDebug } from "@midscene/shared/logger";
import { PROXY_ENDPOINT_FILE, PROXY_PID_FILE, PROXY_UPSTREAM_FILE } from "./cdp-proxy-constants.mjs";
import { cleanupTargetIdFile } from "./cdp-target-store.mjs";
const debug = getDebug('mcp:cdp:proxy');
const PROXY_TERM_GRACE_MS = 2000;
const PROXY_TERM_POLL_MS = 50;
const PROXY_STDERR_BUFFER_LIMIT = 8192;
const PROXY_STARTUP_TIMEOUT_MS = 10000;
function isPageLevelEndpoint(endpoint) {
    return /\/devtools\/page\//.test(endpoint);
}
function resolveBrowserEndpoint(pageEndpoint) {
    return new Promise((resolve, reject)=>{
        let host;
        try {
            const url = new URL(pageEndpoint);
            host = url.host;
        } catch  {
            reject(new Error(`Invalid CDP endpoint URL: ${pageEndpoint}`));
            return;
        }
        const req = node_http.get(`http://${host}/json/version`, {
            timeout: 5000
        }, (res)=>{
            if (res.statusCode && res.statusCode >= 400) {
                reject(new Error(`/json/version returned HTTP ${res.statusCode}`));
                res.resume();
                return;
            }
            let data = '';
            res.on('data', (chunk)=>{
                data += chunk.toString();
            });
            res.on('end', ()=>{
                try {
                    const info = JSON.parse(data);
                    if (info.webSocketDebuggerUrl) resolve(info.webSocketDebuggerUrl);
                    else reject(new Error('webSocketDebuggerUrl not found in /json/version response'));
                } catch  {
                    reject(new Error(`Failed to parse /json/version response: ${data}`));
                }
            });
        });
        req.on('error', (err)=>reject(new Error(`Failed to fetch /json/version: ${err.message}`)));
        req.on('timeout', ()=>{
            req.destroy();
            reject(new Error('Timeout fetching /json/version'));
        });
    });
}
function isProxyAlive() {
    if (!existsSync(PROXY_PID_FILE)) return false;
    try {
        const pid = Number(readFileSync(PROXY_PID_FILE, 'utf-8').trim());
        process.kill(pid, 0);
        return true;
    } catch  {
        return false;
    }
}
function readProxyEndpoint() {
    if (!existsSync(PROXY_ENDPOINT_FILE)) return null;
    try {
        return readFileSync(PROXY_ENDPOINT_FILE, 'utf-8').trim();
    } catch  {
        return null;
    }
}
function readProxyUpstream() {
    if (!existsSync(PROXY_UPSTREAM_FILE)) return null;
    try {
        return readFileSync(PROXY_UPSTREAM_FILE, 'utf-8').trim();
    } catch  {
        return null;
    }
}
function sweepProxyMetadataFiles() {
    try {
        if (existsSync(PROXY_ENDPOINT_FILE)) unlinkSync(PROXY_ENDPOINT_FILE);
    } catch  {}
    try {
        if (existsSync(PROXY_PID_FILE)) unlinkSync(PROXY_PID_FILE);
    } catch  {}
    try {
        if (existsSync(PROXY_UPSTREAM_FILE)) unlinkSync(PROXY_UPSTREAM_FILE);
    } catch  {}
}
async function killProxy() {
    cleanupTargetIdFile();
    if (!existsSync(PROXY_PID_FILE)) return;
    let pid;
    try {
        pid = Number(readFileSync(PROXY_PID_FILE, 'utf-8').trim());
        if (!Number.isFinite(pid)) return void sweepProxyMetadataFiles();
    } catch (err) {
        debug('killProxy: cannot read pid file: %s', err);
        sweepProxyMetadataFiles();
        return;
    }
    try {
        process.kill(pid, 'SIGTERM');
        debug('Sent SIGTERM to proxy pid %d', pid);
    } catch (err) {
        debug('killProxy: SIGTERM failed (pid %d): %s', pid, err);
        sweepProxyMetadataFiles();
        return;
    }
    const deadline = Date.now() + PROXY_TERM_GRACE_MS;
    while(Date.now() < deadline && existsSync(PROXY_PID_FILE))await new Promise((r)=>setTimeout(r, PROXY_TERM_POLL_MS));
    if (existsSync(PROXY_PID_FILE)) {
        debug('proxy pid %d did not clean up within %dms, forcing SIGKILL', pid, PROXY_TERM_GRACE_MS);
        try {
            process.kill(pid, 'SIGKILL');
        } catch  {}
        sweepProxyMetadataFiles();
    }
}
function spawnProxy(chromeEndpoint) {
    return new Promise((resolve, reject)=>{
        const proxyScript = join(__dirname, 'cdp-proxy.js');
        const proc = spawn(process.execPath, [
            proxyScript,
            chromeEndpoint
        ], {
            detached: true,
            stdio: [
                'ignore',
                'pipe',
                'pipe'
            ]
        });
        proc.unref();
        let output = '';
        let stderrBuf = '';
        let settled = false;
        const appendStderr = (chunk)=>{
            stderrBuf += chunk.toString();
            if (stderrBuf.length > PROXY_STDERR_BUFFER_LIMIT) stderrBuf = stderrBuf.slice(-PROXY_STDERR_BUFFER_LIMIT);
        };
        proc.stderr.on('data', appendStderr);
        const formatStderr = ()=>{
            const trimmed = stderrBuf.trim();
            return trimmed ? ` (stderr: ${trimmed})` : '';
        };
        const timer = setTimeout(()=>{
            if (!settled) {
                settled = true;
                reject(new Error(`Proxy startup timeout (${PROXY_STARTUP_TIMEOUT_MS / 1000}s)${formatStderr()}`));
            }
        }, PROXY_STARTUP_TIMEOUT_MS);
        const onData = (chunk)=>{
            output += chunk.toString();
            const lines = output.split('\n');
            for (const line of lines)if (line.trim()) try {
                const parsed = JSON.parse(line);
                if (parsed.endpoint && !settled) {
                    settled = true;
                    clearTimeout(timer);
                    proc.stdout.removeListener('data', onData);
                    proc.stderr.removeListener('data', appendStderr);
                    proc.stdout.destroy();
                    proc.stderr.destroy();
                    resolve(parsed.endpoint);
                    return;
                }
            } catch  {}
        };
        proc.stdout.on('data', onData);
        proc.on('error', (err)=>{
            if (!settled) {
                settled = true;
                clearTimeout(timer);
                reject(new Error(`Failed to spawn proxy: ${err.message}`));
            }
        });
        proc.on('exit', (code, signal)=>{
            if (!settled) {
                settled = true;
                clearTimeout(timer);
                const how = signal ? `signal ${signal}` : `code ${code}`;
                reject(new Error(`Proxy exited with ${how} before ready${formatStderr()}`));
            }
        });
    });
}
async function getProxyEndpoint(chromeEndpoint) {
    let browserEndpoint = chromeEndpoint;
    if (isPageLevelEndpoint(chromeEndpoint)) {
        debug('Page-level CDP endpoint detected, resolving via /json/version: %s', chromeEndpoint);
        try {
            browserEndpoint = await resolveBrowserEndpoint(chromeEndpoint);
            debug('Resolved browser endpoint: %s', browserEndpoint);
        } catch (err) {
            throw new Error(`Cannot use page-level CDP endpoint directly. Puppeteer requires a browser-level endpoint (e.g., ws://host:port/devtools/browser/<id>). Auto-resolution via /json/version failed: ${err.message}. Please provide a browser-level CDP endpoint instead.`);
        }
    }
    if (isProxyAlive()) {
        const endpoint = readProxyEndpoint();
        const savedUpstream = readProxyUpstream();
        if (endpoint) if (!savedUpstream || savedUpstream === browserEndpoint) return endpoint;
        else {
            debug('Proxy connected to different upstream (%s), killing', savedUpstream);
            await killProxy();
        }
    }
    try {
        return await spawnProxy(browserEndpoint);
    } catch (err) {
        console.warn(`[cdp] proxy failed, falling back to direct connection: ${err}`);
        return browserEndpoint;
    }
}
export { getProxyEndpoint, isProxyAlive, killProxy, readProxyEndpoint, readProxyUpstream };
