// @midscene/web/dist/es/mcp-tools.mjs
import { ScreenshotItem, z } from "@midscene/core";
import { BaseMidsceneTools } from "@midscene/shared/mcp/base-tools";
import { AgentOverChromeBridge } from "./bridge-mode/index.mjs";
import { StaticPage } from "./static/index.mjs";
class WebMidsceneTools extends BaseMidsceneTools {
    createTemporaryDevice() {
        return new StaticPage({
            screenshot: ScreenshotItem.create('', Date.now()),
            shotSize: {
                width: 1920,
                height: 1080
            },
            shrunkShotToLogicalRatio: 1
        });
    }
    async ensureAgent(openNewTabWithUrl) {
        if (this.agent && openNewTabWithUrl) {
            try {
                await this.agent?.destroy?.();
            } catch (error) {
                console.debug('Failed to destroy agent during re-init:', error);
            }
            this.agent = void 0;
        }
        if (this.agent) return this.agent;
        this.agent = await this.initBridgeModeAgent(openNewTabWithUrl);
        return this.agent;
    }
    async initBridgeModeAgent(url) {
        const agent = new AgentOverChromeBridge({
            closeConflictServer: true
        });
        if (url) await agent.connectNewTabWithUrl(url);
        else await agent.connectCurrentTab();
        return agent;
    }
    preparePlatformTools() {
        return [
            {
                name: 'web_connect',
                description: 'Connect to web page. If URL provided, opens new tab; otherwise connects to current tab.',
                schema: {
                    url: z.string().url().optional().describe('URL to open in new tab (omit to connect current tab)')
                },
                handler: async (args)=>{
                    const { url } = args;
                    if (this.agent) {
                        try {
                            await this.agent.destroy?.();
                        } catch  {}
                        this.agent = void 0;
                    }
                    this.agent = await this.initBridgeModeAgent(url);
                    const screenshot = await this.agent.page?.screenshotBase64();
                    const label = url ?? 'current tab';
                    return {
                        content: [
                            {
                                type: 'text',
                                text: `Connected to: ${label}`
                            },
                            ...screenshot ? this.buildScreenshotContent(screenshot) : []
                        ]
                    };
                }
            },
            {
                name: 'web_disconnect',
                description: 'Disconnect from current web page and release browser resources',
                schema: {},
                handler: this.createDisconnectHandler('web page')
            }
        ];
    }
}
export { WebMidsceneTools };
