// @midscene/web/dist/es/mcp-server.mjs
import { BaseMCPServer, createMCPServerLauncher } from "@midscene/shared/mcp";
import { WebMidsceneTools } from "./mcp-tools.mjs";
class WebMCPServer extends BaseMCPServer {
    createToolsManager() {
        return new WebMidsceneTools();
    }
    constructor(toolsManager){
        super({
            name: '@midscene/web-bridge-mcp',
            version: "1.7.5",
            description: 'Control the browser using natural language commands'
        }, toolsManager);
    }
}
function mcpServerForAgent(agent) {
    return createMCPServerLauncher({
        agent,
        platformName: 'Web',
        ToolsManagerClass: WebMidsceneTools,
        MCPServerClass: WebMCPServer
    });
}
async function mcpKitForAgent(agent) {
    const toolsManager = new WebMidsceneTools();
    const webAgent = agent;
    toolsManager.setAgent(webAgent);
    await toolsManager.initTools();
    return {
        description: 'Midscene Bridge MCP Server: Control the browser using natural language commands for navigation, clicking, input, hovering, screenshots waitFor, and achieving goals.',
        tools: toolsManager.getToolDefinitions()
    };
}
export { WebMCPServer, mcpKitForAgent, mcpServerForAgent };
