// @midscene/web/dist/es/mcp-tools-puppeteer.mjs
import { spawn } from "node:child_process";
import { existsSync } from "node:fs";
import { mkdir, readFile, unlink, writeFile } from "node:fs/promises";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { ScreenshotItem, z } from "@midscene/core";
import { BaseMidsceneTools } from "@midscene/shared/mcp/base-tools";
import { resolveChromePath } from "@midscene/shared/mcp/chrome-path";
import puppeteer_core from "puppeteer-core";
import { PuppeteerAgent } from "./puppeteer/index.mjs";
import { StaticPage } from "./static/index.mjs";
const ENDPOINT_FILE = join(tmpdir(), 'midscene-puppeteer-endpoint');
const USER_DATA_DIR = join(tmpdir(), 'midscene-puppeteer-profile');
const browserManager = {
    activeBrowser: null,
    async getOrLaunch () {
        if (existsSync(ENDPOINT_FILE)) try {
            const endpoint = (await readFile(ENDPOINT_FILE, 'utf-8')).trim();
            const browser = await puppeteer_core.connect({
                browserWSEndpoint: endpoint,
                defaultViewport: null
            });
            return {
                browser,
                reused: true
            };
        } catch  {
            try {
                await unlink(ENDPOINT_FILE);
            } catch  {}
        }
        const wsEndpoint = await this.launchDetachedChrome();
        await writeFile(ENDPOINT_FILE, wsEndpoint);
        const browser = await puppeteer_core.connect({
            browserWSEndpoint: wsEndpoint,
            defaultViewport: null
        });
        return {
            browser,
            reused: false
        };
    },
    async closeBrowser () {
        if (!existsSync(ENDPOINT_FILE)) return;
        try {
            const endpoint = (await readFile(ENDPOINT_FILE, 'utf-8')).trim();
            const browser = await puppeteer_core.connect({
                browserWSEndpoint: endpoint
            });
            await browser.close();
        } catch  {}
        try {
            await unlink(ENDPOINT_FILE);
        } catch  {}
    },
    disconnect () {
        if (this.activeBrowser) {
            this.activeBrowser.disconnect();
            this.activeBrowser = null;
        }
    },
    async launchDetachedChrome () {
        const chromePath = resolveChromePath();
        await mkdir(USER_DATA_DIR, {
            recursive: true
        });
        const args = [
            '--headless=new',
            `--user-data-dir=${USER_DATA_DIR}`,
            '--remote-debugging-port=0',
            '--no-first-run',
            '--no-default-browser-check',
            '--disable-extensions',
            '--disable-default-apps',
            '--disable-sync',
            '--disable-background-networking',
            '--password-store=basic',
            '--use-mock-keychain',
            '--window-size=1280,800',
            '--force-color-profile=srgb'
        ];
        const proc = spawn(chromePath, args, {
            detached: true,
            stdio: [
                'ignore',
                'ignore',
                'pipe'
            ]
        });
        proc.unref();
        return new Promise((resolve, reject)=>{
            let output = '';
            const onData = (data)=>{
                output += data.toString();
                const match = output.match(/DevTools listening on (ws:\/\/[^\s]+)/);
                if (match) {
                    proc.stderr.removeListener('data', onData);
                    resolve(match[1]);
                }
            };
            proc.stderr.on('data', onData);
            proc.on('exit', (code)=>{
                proc.stderr.removeListener('data', onData);
                reject(new Error(`Chrome exited with code ${code} before DevTools was ready.\nChrome stderr: ${output}\nTip: try setting MIDSCENE_MCP_NO_SANDBOX=1 if running in a container.`));
            });
            setTimeout(()=>reject(new Error(`Chrome launch timeout.\nChrome stderr: ${output}\nTip: try setting MIDSCENE_MCP_NO_SANDBOX=1 if running in a container.`)), 15000);
        });
    }
};
class WebPuppeteerMidsceneTools extends BaseMidsceneTools {
    createTemporaryDevice() {
        return new StaticPage({
            screenshot: ScreenshotItem.create('', Date.now()),
            shotSize: {
                width: 1920,
                height: 1080
            },
            shrunkShotToLogicalRatio: 1
        });
    }
    async ensureAgent(navigateToUrl) {
        if (this.agent && navigateToUrl) {
            try {
                await this.agent?.destroy?.();
            } catch  {}
            this.agent = void 0;
        }
        if (this.agent) return this.agent;
        const { browser, reused } = await browserManager.getOrLaunch();
        browserManager.activeBrowser = browser;
        const pages = await browser.pages();
        let page;
        if (navigateToUrl) {
            page = await browser.newPage();
            await page.goto(navigateToUrl, {
                timeout: 30000,
                waitUntil: 'domcontentloaded'
            });
        } else {
            const webPages = pages.filter((p)=>/^https?:\/\//.test(p.url()));
            page = webPages.length > 0 ? webPages[webPages.length - 1] : pages[pages.length - 1] || await browser.newPage();
            if (reused) await page.bringToFront();
        }
        this.agent = new PuppeteerAgent(page);
        return this.agent;
    }
    async destroy() {
        await super.destroy();
        browserManager.disconnect();
    }
    preparePlatformTools() {
        return [
            {
                name: 'web_connect',
                description: 'Connect to a web page. Opens a new tab with the given URL, or reuses the current page.',
                schema: {
                    url: z.string().url().optional().describe('URL to open in new tab (omit to use current page)')
                },
                handler: async (args)=>{
                    const { url } = args;
                    if (this.agent) {
                        try {
                            await this.agent.destroy?.();
                        } catch  {}
                        this.agent = void 0;
                    }
                    this.agent = await this.ensureAgent(url);
                    const screenshot = await this.agent.page?.screenshotBase64();
                    const label = url ?? 'current page';
                    return {
                        content: [
                            {
                                type: 'text',
                                text: `Connected to: ${label}`
                            },
                            ...screenshot ? this.buildScreenshotContent(screenshot) : []
                        ]
                    };
                }
            },
            {
                name: 'web_disconnect',
                description: 'Disconnect from current web page. The browser stays running for future calls.',
                schema: {},
                handler: async ()=>{
                    if (this.agent) {
                        try {
                            await this.agent.destroy?.();
                        } catch  {}
                        this.agent = void 0;
                    }
                    browserManager.disconnect();
                    return this.buildTextResult('Disconnected from web page (browser still running)');
                }
            },
            {
                name: 'web_close',
                description: 'Close the browser completely and release all resources.',
                schema: {},
                handler: async ()=>{
                    if (this.agent) {
                        try {
                            await this.agent.destroy?.();
                        } catch  {}
                        this.agent = void 0;
                    }
                    await browserManager.closeBrowser();
                    return this.buildTextResult('Browser closed');
                }
            }
        ];
    }
}
export { WebPuppeteerMidsceneTools };
