// @midscene/web/dist/es/platform.mjs
import { ScreenshotItem } from "@midscene/core";
import { createScreenshotPreviewDescriptor, definePlaygroundPlatform } from "@midscene/playground";
import { StaticPage, StaticPageAgent } from "./static/index.mjs";
function createDefaultWebAgent() {
    const page = new StaticPage({
        shotSize: {
            width: 800,
            height: 600
        },
        screenshot: ScreenshotItem.create('', Date.now()),
        shrunkShotToLogicalRatio: 1
    });
    return new StaticPageAgent(page);
}
const webPlaygroundPlatform = definePlaygroundPlatform({
    id: 'web',
    title: 'Midscene Web Playground',
    description: "Web playground platform descriptor",
    async prepare (options) {
        const agent = options?.agent;
        const agentFactory = options?.agentFactory || (options?.agent || options?.agentFactory ? void 0 : ()=>createDefaultWebAgent());
        return {
            platformId: 'web',
            title: options?.title || 'Midscene Web Playground',
            agent: agent || (agentFactory ? void 0 : createDefaultWebAgent()),
            agentFactory,
            launchOptions: options?.launchOptions,
            preview: options?.preview || createScreenshotPreviewDescriptor({
                title: 'Web page preview'
            }),
            metadata: {
                interfaceType: agent?.interface?.interfaceType || 'web'
            }
        };
    }
});
export { webPlaygroundPlatform };
